---
description: åˆ˜å°æ³½å†™äº2020.7.19
---

# 4.2 å®æˆ˜äºŒ | STRT-Seq | å°é¼ å¤§è„‘

## 1 å‰è¨€

ä½¿ç”¨äº†ä¸€ä¸ªå­˜åœ¨å¼‚è´¨æ€§çš„æ•°æ®é›†ï¼Œæ˜¯ç ”ç©¶å°é¼ å¤§è„‘çš„ ([Zeisel et al. 2015](https://pubmed.ncbi.nlm.nih.gov/25700174/))

å…¶ä¸­å¤§çº¦åŒ…å«3000ä¸ªç»†èƒï¼ŒåŒ…æ‹¬å°‘çªèƒ¶è´¨ç»†èƒï¼Œå°èƒ¶è´¨ç»†èƒå’Œç¥ç»å…ƒç­‰ï¼Œä½¿ç”¨çš„ç»†èƒåˆ†ç¦»å¹³å°æ˜¯Fluidigm C1å¾®æµæ§ç³»ç»Ÿï¼Œå±äºæ¯”è¾ƒæ—©æœŸçš„ç³»ç»Ÿã€[å•ç»†èƒæµ‹åºçš„çŸ¥è¯†](https://www.jianshu.com/p/9fc521bf82ba)ã€‘

![å‘å±•å†å²](<../.gitbook/assets/image (3).png>)

### æ–‡åº“åˆ¶å¤‡æ—¶åŠ å…¥äº†UMI

> **UMIç®€å•è§£é‡Šï¼š** UMIå°±æ˜¯ä¸ºäº†å»é™¤PCRæ‰©å¢åå·®çš„ã€‚ä¸€èˆ¬ä¸€ä¸ªåŸºå› å¯¹åº”å¤šä¸ªUMIæ—¶ï¼Œå‡ºç°å¤šä¸ªreadså«æœ‰åŒä¸€ä¸ªUMIæ—¶ï¼Œè¿™é‡Œåªè®¡æ•°ä¸€æ¬¡ã€‚
>
> **UMIè‹±æ–‡è§£é‡Šï¼š** Each transcript molecule can only produce one UMI count but can yield many reads after fragmentation
>
> **UMIè¯¦ç»†è§£é‡Šï¼š** ä¸ç®¡æ˜¯bulk RNAè¿˜æ˜¯scRNAï¼Œéƒ½éœ€è¦è¿›è¡ŒPCRæ‰©å¢ï¼Œä½†æ˜¯ä¸å¯é¿å…æœ‰ä¸€äº›è½¬å½•æœ¬ä¼šè¢«æ‰©å¢å¤ªå¤šæ¬¡ï¼Œè¶…è¿‡äº†çœŸå®è¡¨è¾¾é‡ã€‚å½“èµ·å§‹æ–‡åº“å¾ˆå°æ—¶ï¼ˆæ¯”å¦‚å•ç»†èƒæ•°æ®ï¼‰ï¼Œå°±éœ€è¦æ›´å¤šæ¬¡çš„PCRè¿‡ç¨‹ï¼Œè¿™ä¸ªæ¬¡æ•°è¶Šå¤šï¼Œå¼•å…¥çš„è¯¯å·®å°±è¶Šå¤§ã€‚UMIå°±æ˜¯Unique Molecular Identifierï¼Œç”±4-10ä¸ªéšæœºæ ¸è‹·é…¸ç»„æˆï¼Œåœ¨mRNAåè½¬å½•åï¼Œè¿›å…¥åˆ°æ–‡åº“ä¸­ï¼Œæ¯ä¸€ä¸ªmRNAéšæœºè¿ä¸Šä¸€ä¸ªUMIï¼Œæ ¹æ®PCRç»“æœå¯ä»¥è®¡æ•°ä¸åŒçš„UMIï¼Œæœ€ç»ˆç»Ÿè®¡mRNAçš„æ•°é‡ã€‚
>
> **UMIå›¾ç‰‡è§£é‡Šï¼š**
>
> ![image-20200719150326484](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-070326.png) ![image-20200719150624540](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-070625.png)
>
> **UMIæœ‰å‡ ä¸ªè¦æ±‚ï¼š**
>
> * ä¸èƒ½æ˜¯[å‡èšç‰©](https://zh.wikipedia.org/zh-hans/%E5%9D%87%E8%81%9A%E7%89%A9) ï¼Œå¦‚AAAAAAAAAA
> * ä¸èƒ½æœ‰Nç¢±åŸº
> * ä¸èƒ½åŒ…å«ç¢±åŸºè´¨é‡ä½äº10çš„ç¢±åŸº

## 2 æ•°æ®å‡†å¤‡

```r
# è‡ªå·±ä¸‹è½½
library(scRNAseq)
sce.zeisel <- ZeiselBrainData()
# æˆ–è€…ä½¿ç”¨ä¹‹å‰åˆ†äº«çš„RData
load('sce.zeisel.RData')

sce.zeisel
# class: SingleCellExperiment 
# dim: 20006 3005 
# metadata(0):
#   assays(1): counts
# rownames(20006): Tspan12 Tshz1 ... mt-Rnr1
# mt-Nd4l
# rowData names(1): featureType
# colnames(3005): 1772071015_C02 1772071017_G12
# ... 1772066098_A12 1772058148_F03
# colData names(10): tissue group # ...
# level1class level2class
# reducedDimNames(0):
#   altExpNames(2): ERCC repeat
```

çœ‹åˆ°è¿™ä¹ˆå‡ ä¸ªä¿¡æ¯ï¼š2ä¸‡å¤šåŸºå› ï¼Œ3005ä¸ªæ ·æœ¬ï¼›åªæœ‰åŸå§‹countçŸ©é˜µï¼›ä½¿ç”¨äº†symbol IDï¼›åŠ å…¥äº†ERCC

```r
# æœ‰57ä¸ªERCC
> dim(altExp(sce.zeisel,'ERCC'))
[1]   57 3005

# è€Œä¸”å·²ç»æ ‡æ³¨äº†çº¿ç²’ä½“åŸºå› 
> table(rowData(sce.zeisel))

endogenous       mito 
     19972         34
```

### **ä¸€ä¸ªé‡è¦æ“ä½œï¼šaggregateAcrossFeatures**

> è‹±æ–‡è§£é‡Šæ˜¯ï¼šSum together expression values (by default, counts) for each feature set in each cell. ä½†æ˜¯åªçœ‹è¯´æ˜è¿˜æ˜¯ä¸å¥½ç†è§£ï¼Œä¸¾ä¸ªä¾‹å­ï¼š

**å¯ä»¥çœ‹åˆ°ä¸‹é¢ä¼šæœ‰å¾ˆå¤šåŸºå› å…·æœ‰å¤šä¸ªloc**

æ¯”å¦‚`OTTMUSG00000016609_loc4`ã€`OTTMUSG00000016609_loc3` å…¶å®å¯ä»¥ç®—ä½œä¸€ä¸ªåŸºå› 

```r
head(rownames(sce.zeisel)[grep("_loc[0-9]+$",rownames(sce.zeisel))])
# [1] "Syne1_loc2"              "Hist1h2ap_loc1"         
# [3] "Inadl_loc1"              "OTTMUSG00000016609_loc4"
# [5] "OTTMUSG00000016609_loc3" "Gm5643_loc2" 

# è¿™æ ·çš„æœ‰300å¤šä¸ª
> length(grep("_loc[0-9]+$",rownames(sce.zeisel)))
[1] 330
```

**å¦‚æœæ‹¿ä¸€ä¸ªåŸºå› æ¥çœ‹**

Syne1æœ‰Syne1\_loc1å’ŒSyne1\_loc2

```r
> length(grep("Syne1",rownames(sce.zeisel)))
[1] 2

counts(sce.zeisel)[grep("Syne1",rownames(sce.zeisel)),][1:2,1:3]
# 1772071015_C02 1772071017_G12 1772071017_A05
# Syne1_loc2             11              2              4
# Syne1_loc1              0              0              4
```

**å¦‚æœä½¿ç”¨è¿™ä¸ªå‡½æ•°ï¼Œä¼šæœ‰æ€æ ·æ•ˆæœ**

```r
test <- aggregateAcrossFeatures(sce.zeisel, 
                                      id=sub("_loc[0-9]+$", "", rownames(sce.zeisel)))
# åªå‰©ä¸€ä¸ªäº†ï¼Œä¹Ÿå°±æ˜¯åˆäºŒä¸ºä¸€
> length(grep("Syne1",rownames(test)))
[1] 1

# çœ‹è¡¨è¾¾é‡ï¼Œä¹Ÿæ˜¯åˆäºŒä¸ºä¸€
> counts(test)[grep("Syne1",rownames(test)),][1:3]
1772071015_C02 1772071017_G12 1772071017_A05 
            11              2              8
```

å› æ­¤ï¼Œæ˜ç™½äº†ï¼Œ**è¿™ä¸ªå‡½æ•°å°±æ˜¯å¤„ç†ç›¸åŒè¡Œï¼šæŠŠå‡ ä¸ªç›¸åŒçš„è¡Œçš„å€¼åŠ åœ¨ä¸€èµ·å˜ä¸ºä¸€è¡Œ**

ä¹Ÿå°±æ˜ç™½äº†ï¼Œä¸‹é¢ğŸ‘‡ä¸ºä»€ä¹ˆè¦è¿›è¡Œ`sub`æ“ä½œï¼Œå…¶å®å°±æ˜¯ä¸ºäº†æŠŠlocå»æ‰ï¼Œæš´éœ²å‡ºç›¸åŒçš„åŸºå› åï¼Œæ‰èƒ½æ‰§è¡Œ`aggregateAcrossFeatures`å‡½æ•°

```r
sce.zeisel <- aggregateAcrossFeatures(sce.zeisel, 
                                      id=sub("_loc[0-9]+$", "", rownames(sce.zeisel)))

> dim(sce.zeisel)
[1] 19839  3005
```

### **å†æ·»åŠ Ensembl ID**

```r
library(org.Mm.eg.db)
rowData(sce.zeisel)$Ensembl <- mapIds(org.Mm.eg.db, 
    keys=rownames(sce.zeisel), keytype="SYMBOL", column="ENSEMBL")
```

## 3 è´¨æ§

### å¤‡ä»½æ•°æ®

è¿˜æ˜¯å¤‡ä»½ä¸€ä¸‹ï¼ŒæŠŠunfilteredæ•°æ®ä¸»è¦ç”¨åœ¨è´¨æ§çš„æ¢ç´¢ä¸Š

```r
unfiltered <- sce.zeisel
```

è¿™ä¸ªå…¬å…±æ•°æ®çš„ä½œè€…åœ¨å‘è¡¨æ–‡ç« æ—¶å°†æ•°æ®çš„ä½è´¨é‡ç»†èƒå»æ‰äº†ï¼Œä½†å¹¶ä¸å¦¨ç¢æˆ‘ä»¬åšä¸ªè´¨æ§ï¼Œä¹Ÿå¯ä»¥çœ‹çœ‹å®ƒå»é™¤çš„æ€æ ·

```r
stats <- perCellQCMetrics(sce.zeisel, subsets=list(
    Mt=rowData(sce.zeisel)$featureType=="mito"))
qc <- quickPerCellQC(stats, percent_subsets=c("altexps_ERCC_percent", 
    "subsets_Mt_percent"))
sce.zeisel <- sce.zeisel[,!qc$discard]

> sum(qc$discard)
[1] 189

> dim(sce.zeisel)
[1] 19839  2816
```

### **æ ¹æ®åŸæ¥çš„æ•°æ®ï¼ŒåŠ ä¸Šè´¨æ§æ ‡å‡†ä½œå›¾**

```r
colData(unfiltered) <- cbind(colData(unfiltered), stats)
unfiltered$discard <- qc$discard
# åšä¸ªå›¾
gridExtra::grid.arrange(
    plotColData(unfiltered, y="sum", colour_by="discard") +
        scale_y_log10() + ggtitle("Total count"),
    plotColData(unfiltered, y="detected", colour_by="discard") +
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(unfiltered, y="altexps_ERCC_percent",
        colour_by="discard") + ggtitle("ERCC percent"),
    plotColData(unfiltered, y="subsets_Mt_percent",
        colour_by="discard") + ggtitle("Mito percent"),
    ncol=2
)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-085835.png)

### **å†çœ‹ä¸‹æ–‡åº“å¤§å°å’ŒERCCåˆ†åˆ«å’Œçº¿ç²’ä½“å«é‡çš„å…³ç³»**

```r
gridExtra::grid.arrange(
    plotColData(unfiltered, x="sum", y="subsets_Mt_percent",
        colour_by="discard") + scale_x_log10(),
    plotColData(unfiltered, x="altexps_ERCC_percent", y="subsets_Mt_percent",
        colour_by="discard"),
    ncol=2
)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-085923.png)

### **ç„¶åæ£€æŸ¥ä¸€ä¸‹è¢«è¿‡æ»¤çš„åŸå› **

```r
##              low_lib_size            low_n_features high_altexps_ERCC_percent 
##                         0                         3                        65 
##   high_subsets_Mt_percent                   discard 
##                       128                       189
```

## 4 å½’ä¸€åŒ–

è¿™é‡Œç»†èƒæ•°é‡è¾ƒå¤šï¼Œå› æ­¤éœ€è¦é¢„å…ˆåˆ†ç¾¤+å»å·ç§¯è®¡ç®—size factor

```r
library(scran)
set.seed(1000)
clusters <- quickCluster(sce.zeisel)
sce.zeisel <- computeSumFactors(sce.zeisel, cluster=clusters) 
sce.zeisel <- logNormCounts(sce.zeisel)
summary(sizeFactors(sce.zeisel))
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   0.119   0.486   0.831   1.000   1.321   4.509
```

**çœ‹çœ‹ä¸¤ç§å½’ä¸€åŒ–æ–¹æ³•çš„å·®å¼‚**

```r
# å¸¸è§„ï¼šæœ€ç®€å•çš„åªè€ƒè™‘æ–‡åº“å¤§å°
summary(librarySizeFactors(sce.zeisel))
# Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
# 0.1757  0.5680  0.8680  1.0000  1.2783  4.0839 

plot(librarySizeFactors(sce.zeisel), sizeFactors(sce.zeisel), pch=16,
    xlab="Library size factors", ylab="Deconvolution factors", log="xy")
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-090527.png)

## 5 æ‰¾è¡¨è¾¾é‡é«˜å˜åŒ–åŸºå› 

ç†è®ºä¸Šï¼Œåº”è¯¥å¯¹æ¯ä¸ªç»†èƒéƒ½æ ‡è®°æ‰¹æ¬¡ä¿¡æ¯ï¼Œæ·»åŠ `block`ä¿¡æ¯ã€‚ä½†æ˜¯è¿™é‡Œç”±äºæŠ€æœ¯ä¸åŒï¼Œæ¯ä¸ªæ¿å­ä¸Šåªæœ‰20-40ä¸ªç»†èƒï¼Œå¹¶ä¸”ç»†èƒç¾¤ä½“å…·æœ‰é«˜åº¦å¼‚è´¨æ€§ï¼Œä¸èƒ½å‡è®¾æ¯ä¸ªæ¿ä¸Šçš„ç»†èƒç±»å‹çš„åˆ†å¸ƒæ˜¯ç›¸åŒçš„ï¼Œå› æ­¤è¿™é‡Œä¸ä½¿ç”¨`block`å°†æ‰¹æ¬¡ä¿¡æ¯â€œé”ä½â€æ˜¯åˆé€‚çš„

æ—¢ç„¶æœ‰ERCCï¼Œå°±å¯ä»¥ç”¨ç¬¬ä¸‰ç§æ–¹æ³•ã€åœ¨ä¹‹å‰ [3.3 æŒ‘é€‰é«˜å˜åŒ–åŸºå› ](https://jieandze1314.osca.top/03/03-3) çš„2.3 è€ƒè™‘æŠ€æœ¯å™ªéŸ³ã€‘ï¼š

```r
dec.zeisel <- modelGeneVarWithSpikes(sce.zeisel, "ERCC")
top.hvgs <- getTopHVGs(dec.zeisel, prop=0.1)
> length(top.hvgs)
[1] 1816
```

åŒæ ·ä½¿ç”¨äº†spike-inï¼Œå¯¹æ¯”ä¸€ä¸‹ï¼Œçœ‹åˆ°è¿™é‡Œä¸ç®¡æ€»ä½“æ–¹å·®è¿˜æ˜¯æŠ€æœ¯å› ç´ æ–¹å·®éƒ½è¦æ¯”ä¹‹å‰smart-seq2è¦å°ã€‚

> smart-seq2æ˜¯æŒ‰readè®¡æ•°ï¼Œè¿™é‡Œç”±äºæ·»åŠ äº†UMIï¼Œæ˜¯æŒ‰moleculeè®¡æ•°ï¼Œä¹Ÿå°±æ˜¯è¯´ï¼Œ**UMIçš„åŠ å…¥ï¼Œç¡®å®å‡å°‘äº†PCRæ‰©å¢çš„åå·®å½±å“**

å¦å¤–å›¾ä¸­çœ‹åˆ°ï¼Œè¿™é‡ŒSTRT-seqçš„spike-inæ–¹å·®ä¸€ç›´è¦æ¯”å†…æºåŸºå› çš„æ–¹å·®å°ï¼Œä¹Ÿå°±æ˜¯è¯´å†…æºåŸºå› çš„å˜åŒ–å¹…åº¦ä¸€ç›´ä¿æŒé«˜ä½ï¼Œä½“ç°äº†æ•°æ®ä¸­åŒ…å«å¤šç§ç»†èƒç±»å‹è€Œå¯¼è‡´çš„å¼‚è´¨æ€§ï¼Œå¼‚è´¨æ€§å¯¼è‡´äº†åŸºå› è¡¨è¾¾æåº¦ä¸å‡è¡¡

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-091219.png)

## 6 é™ç»´

```r
library(BiocSingular)
set.seed(101011001)
sce.zeisel <- denoisePCA(sce.zeisel, technical=dec.zeisel, subset.row=top.hvgs)
sce.zeisel <- runTSNE(sce.zeisel, dimred="PCA")
```

çœ‹ä¸€ä¸‹å¾—åˆ°çš„PCæ•°

```r
ncol(reducedDim(sce.zeisel, "PCA"))
## [1] 50
```

## 7 èšç±»

```r
snn.gr <- buildSNNGraph(sce.zeisel, use.dimred="PCA")
colLabels(sce.zeisel) <- factor(igraph::cluster_walktrap(snn.gr)$membership)
```

çœ‹ä¸€ä¸‹ç»“æœ

```r
table(colLabels(sce.zeisel))
## 
##   1   2   3   4   5   6   7   8   9  10  11  12  13  14 
## 283 451 114 143 599 167 191 128 350  70 199  58  39  24
```

ç”»å›¾

```r
plotTSNE(sce.zeisel, colour_by="label")
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-092215.png)

## 8 æ‰¾markeråŸºå› å¹¶è§£é‡Šç»“æœ

> ä¸»è¦è¿˜æ˜¯å…³æ³¨ä¸Šè°ƒåŸºå› ï¼Œå¯ä»¥å¸®æˆ‘ä»¬å¿«é€Ÿåˆ¤æ–­å‡ºå¼‚è´¨æ€§ç¾¤ä½“ä¸­å„ä¸ªç»†èƒç±»å‹çš„å·®å¼‚

æ¯”å¦‚è¿˜æ˜¯é’ˆå¯¹cluster1çœ‹çœ‹

```r
markers <- findMarkers(sce.zeisel, direction="up")
marker.set <- markers[["1"]]

> ncol(marker.set)
[1] 17

> colnames(marker.set)
 [1] "Top"           "p.value"       "FDR"           "summary.logFC" "logFC.2"       "logFC.3"       "logFC.4"      
 [8] "logFC.5"       "logFC.6"       "logFC.7"       "logFC.8"       "logFC.9"       "logFC.10"      "logFC.11"     
[15] "logFC.12"      "logFC.13"      "logFC.14"  


head(marker.set[,1:8], 10)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-092744.png)

### **æ–¹æ³•ä¸€ï¼šç”»çƒ­å›¾**

**ä½¿ç”¨cluster1çš„Top10åŸºå› ï¼ˆä½†ä¸ä¸€å®šåªæ˜¯10ä¸ªï¼‰**

```r
top.markers <- rownames(marker.set)[marker.set$Top <= 10]
> length(top.markers)
[1] 58

plotHeatmap(sce.zeisel, features=top.markers, order_columns_by="label")
```

æ¥ä¸‹æ¥å°±æ˜¯æ ¹æ®èƒŒæ™¯çŸ¥è¯†äº†ï¼Œæ¯”å¦‚çœ‹åˆ°Gad1ã€Slc6a1è¡¨è¾¾é‡éƒ½å¾ˆé«˜ï¼Œå¯èƒ½è¡¨æ˜cluster1å±äºä¸­é—´ç¥ç»å…ƒ

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-093013.png)

### **æ–¹æ³•äºŒï¼šåŸºäºlogFC**

æ¯”å¦‚å¯ä»¥æŒ‘å‡ºcluster1çš„è®¡ç®—ç»“æœ`marker.set`ä¸­å‰50ä¸ªåŸºå› ï¼ˆè¿™é‡Œå°±æ˜¯50ä¸ªï¼Œè€Œä¸æ˜¯Top50ï¼‰ï¼Œç„¶åæ ¹æ®cluster1ä¸å…¶ä»–clustersçš„logFCï¼Œå¯¹æ¯ä¸ªåŸºå› è¡¨è¾¾é‡åšçƒ­å›¾

```r
library(pheatmap)
logFCs <- getMarkerEffects(marker.set[1:50,])
pheatmap(logFCs, breaks=seq(-5, 5, length.out=101))
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-094211.png)

é‚£ä¹ˆ**è¿™ä¸ªå‡½æ•°åˆ°åº•åšäº†ä»€ä¹ˆå‘¢ï¼Ÿ**çœ‹å›¾å°±çŸ¥é“ï¼š

> ä¹Ÿå°±æ˜¯æŠŠæ¯ä¸ªåŸºå› åœ¨å…¶ä»–clustersçš„logFCç»“æœæŒ‘å‡ºæ¥ï¼Œæ±‡é›†æˆäº†ä¸€ä¸ªæ–°çŸ©é˜µï¼Œæˆ‘ä»¬è‡ªå·±æ‰‹åŠ¨ä¹Ÿæ˜¯å¯ä»¥åšåˆ°çš„

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2020-07-19-094036.png)
