---
description: åˆ˜å°æ³½å†™äº19.8.21+28
---

# 5.4.1 Monocle V3å®æˆ˜

> **å­¦ä¹ æ„Ÿæƒ³ï¼š **\
> è¿™æ¬¡èŠ±è´¹çš„æ—¶é—´è¾ƒé•¿ï¼Œä¸ªäººæ„Ÿè§‰è¿™ä¸ªåŒ…æ¯”Seuratã€Scateréš¾å­¦ï¼Œè¯´æ˜æ–‡æ¡£å†™çš„å®åœ¨æ˜¯å¤ªé•¿äº†ï¼

### å‰è¨€

> ä¾æ—§ä¸»æ‰“å…¨äººå·¥ç†è§£ï¼ˆ**ä¸æ˜¯ç¿»è¯‘**ï¼‰

å¯¹è¿™æ¬¾è½¯ä»¶çš„äº†è§£ä¸»è¦æ˜¯å› ä¸ºå®ƒåšçš„å‘è‚²è½¨è¿¹ã€æ‹Ÿæ—¶åºåˆ†æå¾ˆæ¼‚äº®ï¼Œè¿™ä¹Ÿæ˜¯å®ƒçš„ä¸»æ‰“

Monocleçš„å®˜ç½‘åœ¨ï¼š

* ç‰ˆæœ¬2ï¼š[https://cole-trapnell-lab.github.io/monocle-release/docs/#installing-monocle](https://cole-trapnell-lab.github.io/monocle-release/docs/#installing-monocle)
* ç‰ˆæœ¬3ï¼š[https://cole-trapnell-lab.github.io/monocle3/monocle3\_docs/](https://cole-trapnell-lab.github.io/monocle3/monocle3\_docs/)

é¦–å…ˆçœ‹æ–°ç‰ˆæœ¬çš„ç‰¹æ€§å’Œå‡çº§ä¹‹å¤„ï¼š

* å¤„ç†çš„ç»†èƒæ•°å¢åŠ å¾ˆå¤šï¼ˆmillions of cellsï¼‰
* é’ˆå¯¹å‘è‚²ç”Ÿç‰©å­¦é¢†åŸŸï¼Œåšäº†ä¸€äº›é‡å¤§æ”¹è¿›ï¼š
  * å‘è‚²è½¨è¿¹çš„ç ”ç©¶æµç¨‹ä¼˜åŒ–
  * æ”¯æŒ[UMAP](https://github.com/lmcinnes/umap)æ¨æ–­å‘è‚²è½¨è¿¹ï¼Œæ— ç¼è¡”æ¥åˆ°`reduceDimension`å‡½æ•°ï¼Œæˆ–è€…ä½¿ç”¨ç‹¬ç«‹çš„å‡½æ•°`UMAP` 
  * æ”¯æŒå¤šä¸ªç¥–æºï¼ˆtootsï¼‰çš„å‘è‚²æ¨æ–­
  * åˆ©ç”¨è¿™ä¸ªç®—æ³•[approximate graph abstraction](https://www.biorxiv.org/content/early/2017/10/25/208819) ç†è§£å„æ¡å‘è‚²è½¨è¿¹çš„åˆ†ç¦»åŠå¹³è¡Œå‘è‚²çš„è½¨è¿¹
  * 3Dæ„å»ºå‘è‚²è½¨è¿¹
* é™¤äº†å‘è‚²è½¨è¿¹ä»¥å¤–ï¼Œå®˜æ–¹è¿˜ä»‹ç»å®ƒåœ¨äºšå‹å‘ç°ã€å·®å¼‚åˆ†ææ–¹é¢è¡¨ç°ä¸é”™
* å¦å¤–ï¼Œç‰ˆæœ¬3åœ¨æŒç»­ä¸æ–­ä¼˜åŒ–ï¼Œå®˜æ–¹ç§°æ¯å‡ ä¸ªå‘¨å°±ä¼šå¢åŠ ä¸€äº›æ–°åŠŸèƒ½
* å…¶ä¸­çš„ç®—æ³•ç»†èŠ‚çœ‹2019å¹´å‘è¡¨çš„[Cao & Spielmann et al](http://dx.doi.org/10.1038/s41586-019-0969-x). 

å®ƒçš„ä¸»è¦åˆ†ææµç¨‹æ˜¯ï¼š

![Monocleä¸»è¦åˆ†ææµç¨‹](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-22-015711.png)

> **æ³¨æ„ï¼šä»¥ä¸‹æ˜¯ä»‹ç»ç‰ˆæœ¬3ï¼ˆç‰ˆæœ¬2çš„ä½¿ç”¨å’Œ3å­˜åœ¨ä¸€äº›ä¸åŒä¹‹å¤„ï¼Œä¼šåœ¨â€è·Ÿç€å®˜ç½‘å­¦ä¹ â€œè¿™ä¸€éƒ¨åˆ†çš„æœ€åæ ‡æ³¨ï¼‰**

### å®‰è£…ç‰ˆæœ¬3

> è¦æ±‚Rç‰ˆæœ¬3.5ä»¥ä¸Šï¼ŒBioconductorç‰ˆæœ¬3.5ä»¥ä¸Š

```r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
# ä¸€äº›ä¾èµ–åŒ…
bioc_pkgs <- c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
                       'limma', 'S4Vectors', 'SingleCellExperiment',
                       'SummarizedExperiment')
for (bpkg in bioc_pkgs){
  if (! require(bpkg,character.only=T) ) {
    BiocManager::install(bpkg,ask = F,update = F)
    require(bpkg,character.only=T)
    }
}

# (å¯é€‰)å¦‚æœè¦åœ¨ç»†èƒèšç±»æ—¶è®¾å®šåˆ†è¾¨ç‡å‚æ•°(resolution)ï¼Œå°±è¦å®‰è£…ä¸€ä¸ªpythonåŒ…
install.packages("reticulate")
reticulate::py_install("louvain")
# ç°åœ¨å®‰è£…3ç‰ˆæœ¬ï¼Œè¿˜æ˜¯è¦é€šè¿‡github
devtools::install_github('cole-trapnell-lab/monocle3')
# é‡å¯Rstudioï¼Œæ£€æµ‹æ˜¯å¦æˆåŠŸ
library(monocle3)
```

### Monocleç¤ºä¾‹1-ç»†èƒèšç±»åŠé‰´å®šäºšç¾¤

**åˆ›å»ºå¯¹è±¡**

æ•´ä½“ä»£ç å°±æ˜¯è¿™æ ·ï¼ˆå…ˆäº†è§£æ¦‚å†µç„¶åä¸‹é¢æœ‰å®é™…ç»ƒä¹ ï¼‰ï¼š

```r
cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
```

ä½†æ˜¯è¿™ä¸ªå¯¹è±¡`cell_data_set`çš„è®¾ç½®éœ€è¦å¥½å¥½ç†è§£ï¼š

* è¿™ä¸ªå¯¹è±¡çš„çµæ„Ÿæ¥è‡ª[SingleCellExperiment](http://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html)ï¼Œå› æ­¤å®ƒçš„æ“ä½œä¹Ÿæ˜¯ç±»ä¼¼çš„
* `expression_matrix` æ˜¯ä¸€ä¸ªè¡Œä¸ºåŸºå› ï¼Œåˆ—ä¸ºç»†èƒæ ·æœ¬çš„è¡¨è¾¾çŸ©é˜µ
* `cell_metadata` æ˜¯ä¸€ä¸ªæ•°æ®æ¡†ï¼Œè¡Œä¸ºç»†èƒï¼Œåˆ—æ˜¯ç»†èƒçš„å±æ€§ï¼ˆæ¯”å¦‚ç»†èƒç±»å‹ã€åŸ¹å…»ç¯å¢ƒã€åŸ¹å…»æ—¶é—´ç­‰ï¼‰
* `gene_metadata`æ˜¯ä¸€ä¸ªæ•°æ®æ¡†ï¼Œ**è¡Œä¸ºfeatureä¿¡æ¯ï¼ˆæ¯”å¦‚åŸºå› ï¼‰ï¼Œåˆ—æ˜¯åŸºå› å±æ€§ï¼ˆæ¯”å¦‚GCå«é‡ï¼‰**ã€è¿™ä¸ªå¾ˆæ–°é²œï¼Œå› ä¸ºå…¶ä»–çš„åŸºäºSingleCellExperimentçš„åŒ…å¦‚`Scater` åªéœ€è¦è¡¨è¾¾çŸ©é˜µå’Œæ ·æœ¬ä¿¡æ¯å³å¯ã€‘ã€‚æ›´é‡è¦çš„æ˜¯ï¼Œè¿™ä¸ªæ•°æ®æ¡†ä¸­**å¿…é¡»æœ‰è¿™ä¹ˆä¸€åˆ—ï¼š`gene_short_name`ï¼Œå…¶ä¸­ä¿å­˜åŸºå› å**ï¼ˆç®€ç§°æˆ–Symbolæ ‡å‡†åå‡å¯ï¼Œç”¨äºä½œå›¾ï¼‰
* å¹¶ä¸”è¦æ»¡è¶³ï¼š`expression_matrix`çš„è¡Œä¸`gene_metadata`çš„è¡Œå¯¹åº”ï¼›`expression_matrix`çš„åˆ—ä¸`cell_metadata`çš„è¡Œå¯¹åº”

ç„¶åæµ‹è¯•ä¸€ä¸‹ï¼ˆ**ä¸‹è½½é€Ÿåº¦ä¼šå¾ˆæ…¢ï¼**ï¼‰

> å°æç¤ºï¼šä½¿ç”¨æœåŠ¡å™¨`axel`ä¸‹è½½ï¼Œé€Ÿåº¦ä¼šåŠ å¿«ï¼ˆä¸è¿‡è¿™è¦çœ‹æœåŠ¡å™¨çš„ç½‘ç»œé…ç½®ï¼‰ï¼Œå¦‚æœç»™åŠ›çš„è¯ï¼Œä¸‹è½½å‡é€Ÿèƒ½åˆ°120kb/sï¼Œä¸€ä¸¤åˆ†é’Ÿå°±ä¸‹è½½å¥½ å¦‚æœç½‘ç»œçœŸçš„ä¸ç»™åŠ›ä¹Ÿæ²¡å…³ç³»ï¼Œæˆ‘å¸®å¤§å®¶ä¸‹è½½ä¸‹æ¥äº†ï¼Œåœ¨è±†è±†å’ŒèŠ±èŠ±çš„å…¬ä¼—å·â€ç”Ÿä¿¡æ˜Ÿçƒâ€œåå°å›å¤â€monocleâ€œå°±å¯ä»¥å¾—åˆ°è¿™ä¸‰ä¸ªæ–‡ä»¶å•¦ï¼š
>
> ![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-22-024036.png)
>
> ![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-22-032551.png)

```r
# ä¸‹è½½æ•°æ®(å¦‚æœå·²ç»æœ‰äº†ï¼Œå¯ä»¥è·³è¿‡)
expression_matrix <- readRDS(url("http://staff.washington.edu/hpliner/data/cao_l2_expression.rds"))
cell_metadata <- readRDS(url("http://staff.washington.edu/hpliner/data/cao_l2_colData.rds"))
gene_annotation <- readRDS(url("http://staff.washington.edu/hpliner/data/cao_l2_rowData.rds"))

# åŠ è½½ã€æ¢ç´¢æ•°æ®
expression_matrix <- readRDS(file = 'cao_l2_expression.rds')
cell_metadata <- readRDS(file = 'cao_l2_colData.rds')
gene_annotation <- readRDS(file = 'cao_l2_rowData.rds')

dim(expression_matrix)
expression_matrix[1:3,1:3]
cell_metadata[1:3,1:3]
head(gene_annotation)

# åˆ›å»ºCDS(Cell Data Set)å¯¹è±¡
cds <- new_cell_data_set(expression_matrix,
                                   cell_metadata = cell_metadata,
                                   gene_metadata = gene_annotation)
cds
```

**æ•°æ®é¢„å¤„ç†**

è¿™ä¸€æ­¥å°±æ˜¯å‘Šè¯‰Monocleï¼Œä½ æƒ³æ€ä¹ˆå¯¹æ•°æ®è¿›è¡Œå½’ä¸€åŒ–ã€æ ‡å‡†åŒ–ï¼Œåˆæ­¥é™ç»´æ˜¯ä½¿ç”¨PCAï¼ˆé’ˆå¯¹æ ‡å‡†çš„RNA-seqï¼Œéœ€è¦æŒ‡å®šè®¡ç®—çš„ä¸»æˆåˆ†æ•°ï¼‰è¿˜æ˜¯LSAï¼ˆLatent semantic analysisï¼Œé’ˆå¯¹ATAC-seqï¼‰ã€‚è¿™äº›éƒ½æ˜¯ä¸ºåé¢çš„èšç±»æ“ä½œï¼ˆtSNEã€UMAPåšå‡†å¤‡ï¼‰

```r
# æ¯”è¾ƒè€—æ—¶
cds <- preprocess_cds(cds, num_dim = 100) #å¤§çº¦2 mins
# è¿™é‡Œè®¾ç½®è¿™ä¹ˆå¤šä¸»æˆåˆ†ï¼Œå¯èƒ½æ˜¯å› ä¸ºç»†èƒæ•°å¤ªå¤š(4ä¸‡å¤šä¸ª)ï¼Œæˆåˆ†å¤ªå°‘ä¸è¶³ä»¥ä»£è¡¨æ•´ä½“
```

è¿™ä¸ª`preprocess_cds`çœ‹ä¸‹å¸®åŠ©æ–‡æ¡£å°±èƒ½å¾—çŸ¥ï¼š

* é™ç»´æ–¹æ³•`method = c("PCA", "LSI")`ï¼Œé»˜è®¤æ˜¯PCA
* é»˜è®¤ç»´åº¦`num_dim`æ˜¯50
* åˆæ­¥é™ç»´å‰å½’ä¸€åŒ–`norm_method` é»˜è®¤æ˜¯logå¤„ç†
* è®¾ç½®é™ç»´æ–¹æ³•æ˜¯PCAæ—¶ï¼Œè‡ªåŠ¨è¿›è¡Œæ ‡å‡†åŒ–å¤„ç†

æ£€æŸ¥ä¸€ä¸‹ä½¿ç”¨çš„ä¸»æˆåˆ†ï¼ˆPCsï¼‰æ˜¯å¦èƒ½å¤ŸæŠ“å–æœ€ä¸»è¦çš„åŸºå› è¡¨è¾¾å˜åŒ–ä¿¡æ¯

```r
plot_pc_variance_explained(cds)
# å¾ˆåƒSeuratçš„ElbowPlot()
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-22-034739.png)

å¯ä»¥çœ‹åˆ°ï¼Œè¶…è¿‡100ä¸ªä¸»æˆåˆ†åå¯¹æ•´ä½“å˜åŒ–çš„è´¡çŒ®å€¼å°±æ²¡æœ‰å¤ªå¤šäº†ï¼Œè€Œä¸”æ¯å¤šä¸€ä¸ªPCå°±ä¼šå¯¹ä¸‹æ¸¸æ•°æ®åˆ†æé€ æˆå‹åŠ›

**ç»§ç»­é™ç»´åŠå¯è§†åŒ–**

**å…³äºç»´åº¦æœ‰å¿…è¦å¥½å¥½å†äº†è§£ä¸‹ï¼š**

> ï¼ˆè¯¦è§`?reduce_dimension`ï¼‰ **é‚£æˆ‘å°±ç¼–ä¸€ä¸ªâ€é™ç»´â€œå°æ•…äº‹å§** \
> æ¯ä¸ªç»†èƒä¸­éƒ½æœ‰2ä¸‡å¤šåŸºå› ï¼Œå¯ä»¥è¢«å½“åšé«˜ç»´ç©ºé—´ä¸Šçš„ä¸€ä¸ªç‚¹ï¼Œè€Œè¿™é‡Œè¯´çš„**æ¯ä¸ªç»´åº¦å°±æ˜¯å°±ä¸åŒåŸºå› çš„è¡¨è¾¾é‡**ï¼Œä¹Ÿå°±æ˜¯è¯´ï¼Œ**æˆ‘ä»¬ç°åœ¨æœ‰2ä¸‡ä¸ªç»´åº¦ä¸Šçš„4ä¸‡ä¸ªç‚¹**ï¼›å¤„ç†çš„ç»´åº¦è¶Šå¤šï¼Œåé¢çš„è½¨è¿¹æ¨æ–­ä¹Ÿå°±æ˜¯è¶Šéš¾ \
> å¥½åœ¨åŸºå› ä¹‹é—´ä¸æ˜¯æ¯«æ— ç“œè‘›çš„ï¼Œå„ç§åŸºå› ä¹‹é—´æ€»ä¼šæœ‰é”™ç»¼å¤æ‚çš„è”ç³»ï¼Œé€šè¿‡æ’é™¤è¿™ç§â€å†¥å†¥ä¸­çš„ç›¸å…³æ€§â€œå°±ä¼š**æå–å‡ºå®ƒä»¬ä¹‹é—´æœ€ä¸åŒçš„åœ°æ–¹**ï¼Œæ›´èƒ½è¯´æ˜æ•´ä½“çš„ä¸€ä¸ªç‰¹å¾ã€‚äºæ˜¯çº¿æ€§é™ç»´å°±èƒ½å°†2ä¸‡ä¸ªç»´åº¦é™åˆ°è¿™é‡Œçš„100ä¸ªï¼Œé‚£ä¹ˆä¹‹åå†å¤„ç†çš„è¯ï¼Œå¯èƒ½çº¿æ€§é™ç»´ä¹Ÿæœ‰å‹åŠ›ã€‚\
> **â€ä¸“äººä¸“äº‹**ï¼šâ€œ\~äºæ˜¯æ‰¾æ¥äº†æ›´ä¸“ä¸šçš„éçº¿æ€§é™ç»´å·¥å…·tSNEã€UMAPï¼Œå®ƒä»¬æœ€ç»ˆèƒ½å®ç°å†å°†è¿™100ä¸ªæˆåˆ†ç»§ç»­å‹ç¼©ï¼Œæœ€ç»ˆå¾—åˆ°æˆ‘ä»¬èƒ½ç†è§£çš„äºŒç»´æˆ–è€…3ç»´ç©ºé—´ï¼ˆé‚£ä¹ˆå…¶ä¸­åŒ…å«çš„æ ·æœ¬ç‰¹å¾ä¸€å®šæ˜¯**ç²¾åä¸­çš„ç²¾å**ï¼‰ è‹±æ–‡çš„è§£é‡Šå°±æ˜¯ï¼šEach cell can be viewed as a point in a high-dimensional space, where each dimension describes the expression of a different gene.

è¿™é‡Œä¸»è¦é‡‡ç”¨éçº¿æ€§é™ç»´çš„æ–¹æ³•ï¼šä¸€ç§æ˜¯ç›®å‰å¾ˆæµè¡Œçš„tSNEï¼Œå¦ä¸€ç§æ˜¯2018å¹´å‘å¸ƒçš„åŸºäºPythonçš„UMAPï¼ˆUMAPå¤„ç†å¤§é‡ç»†èƒçš„ç®—æ³•æ›´ä¼˜ç§€ï¼›å¦å¤–ä¸t-SNEç›¸æ¯”ï¼ŒUMAPå¯ä»¥ä¿ç•™æ›´å¤šçš„ç»†èƒäºšç¾¤è¿ç»­æ€§ï¼‰

```r
cds <- reduce_dimension(cds, preprocess_method = "PCA",reduction_method = c("UMAP"))
# å¤§çº¦è¿è¡Œ1.5 min
```

å…³äºè¿™ä¸ªå‡½æ•°ï¼š

* å¤šç§ç»§ç»­é™ç»´çš„ç®—æ³•ï¼šå¦‚`"UMAP", "tSNE", "PCA" and "LSI"`ï¼Œä¸è¿‡å‡½æ•°é»˜è®¤ä½¿ç”¨UMAP
*   å…³äºè¿è¡ŒåŠ é€Ÿï¼š åŸºäº`BiocParallel`æ”¯æŒå¤šçº¿ç¨‹è¿è¡Œï¼Œä½¿ç”¨`cores`å®šä¹‰ï¼›æˆ–è€…ä½¿ç”¨`umap.fast_sgd=TRUE`å‚æ•°å¯ä»¥åŠ é€Ÿé™ç»´ ä½†æ˜¯è‡ªå·±åŠ é€Ÿçš„è¯ï¼Œå®ƒä¼šå‹æƒ…æç¤ºä¸€ä¸‹ï¼š

    ```r
    # Note: reduce_dimension will produce slightly different output each time you run it unless you set 'umap.fast_sgd = FALSE' and 'cores = 1'
    ```
* é»˜è®¤å¾—åˆ°ä¸¤ä¸ªç»´åº¦
* éœ€è¦å®šä¹‰ä¸Šä¸€æ­¥`preprocess_method`ä½¿ç”¨çš„ç®—æ³•ï¼Œé»˜è®¤æ˜¯`LSI`

è¿›è¡Œå¯è§†åŒ–ï¼š

```r
plot_cells(cds)
```

![é»˜è®¤çš„umapç»“æœ](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-22-041227.png)

å›¾ä¸­çš„æ¯ä¸ªç‚¹éƒ½ä»£è¡¨`cds`å¯¹è±¡ä¸­çš„ä¸åŒç»†èƒï¼Œå¯ä»¥çœ‹åˆ°ä¸åŒçš„ç»†èƒåˆ†æˆäº†ä¸åŒçš„ç¾¤ï¼Œæœ‰çš„ç¾¤æœ‰å‡ åƒä¸ªç»†èƒï¼Œæœ‰çš„ç¾¤åªæœ‰å°‘æ•°å‡ ä¸ªï¼Œæˆ‘ä»¬è¿™é‡Œç›´æ¥ä½¿ç”¨æµ‹è¯•æ•°æ®ä¸­åšå¥½çš„ç»†èƒæ ‡è®°

```r
# æ•°æ®ä¸­ä¸€å…±æä¾›äº†30ç±»ç»†èƒ(ç»†èƒç±»å‹åŠæ•°é‡è§ï¼š)
table(colData(cds)$"cao_cell_type")
# æ ¹æ®è¿™ä¸ªä¸Šè‰²
plot_cells(cds, color_cells_by="cao_cell_type")
# (å¦å¤–é™¤äº†ç”¨cao_cell_typeç»†èƒç±»å‹ï¼Œè¿˜èƒ½ç”¨ä¸‹é¢çš„ä»»æ„ä¸€ä¸ª)
> colnames(colData(cds))
[1] "plate"         "cao_cluster"   "cao_cell_type" "cao_tissue"    "Size_Factor"
```

å¯ä»¥çœ‹åˆ°è®¸å¤šç»†èƒç±»å‹å†UMAPç»“æœä¸­é çš„å¾ˆè¿‘

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-22-041537.png)

æ ¹æ®åŸºå› è¡¨è¾¾é‡è¿›è¡Œæ˜ å°„ï¼š

```r
plot_cells(cds, genes=c("cpna-2", "egl-21", "ram-2", "inos-1"))
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-23-012402.png)

å¦‚æœæƒ³ä½¿ç”¨tSNEæ–¹æ³•é™ç»´çš„è¯ï¼š

```r
cds <- reduce_dimension(cds, reduction_method="tSNE")
# ç„¶åå¯¹tSNEç»“æœå¯è§†åŒ–
plot_cells(cds, reduction_method="tSNE", color_cells_by="cao_cell_type")
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-23-013132.png)

å…¶å®å¯ä»¥çœ‹åˆ°ï¼ŒtSNEç»“æœçš„å„ä¸ªç»†èƒäºšç¾¤å†…éƒ¨çš„å…³è”æ€§ä¸å¦‚UMAPï¼ˆæ¯”å¦‚è¿™ä¸ª`stem cells`ï¼‰

**æ£€æŸ¥ã€ç§»é™¤æ‰¹æ¬¡æ•ˆåº” => residual_model_formula_strå‚æ•°**

ä¸ºäº†æ›´åŠ çœŸå®åœ°åæ˜ å·®å¼‚åŸºå› çš„æ¥æºæ˜¯ç”Ÿç‰©å­¦å› ç´ ï¼Œè€Œä¸æ˜¯æŠ€æœ¯å› ç´ ï¼ˆæ¯”å¦‚ç»†èƒæ¿æ‰¹æ¬¡ã€ç»†èƒåŸ¹å…»å·®åˆ«ã€æµ‹åºå·®åˆ«ç­‰ï¼‰ï¼Œå°±é¦–å…ˆè¦æ£€æŸ¥æœ‰æ²¡æœ‰æ‰€è°“çš„â€æŠ€æœ¯æ‰¹æ¬¡â€œå¯¹æ•°æ®å·®å¼‚äº§ç”Ÿå½±å“ã€‚

å½“è¿›è¡Œé™ç»´åˆ†ææ—¶ï¼Œåº”è¯¥ç»å¸¸æ£€æŸ¥æ‰¹æ¬¡æ•ˆåº”ï¼Œæœ€å¥½åœ¨`colData(cds)` æ·»åŠ ä¸€åˆ—ï¼Œå…¶ä¸­æ³¨æ˜ç»†èƒçš„æ‰¹æ¬¡ä¿¡æ¯ï¼ˆæ¯”å¦‚æ¥è‡ªå“ªä¸ªç»†èƒæ¿ï¼‰ï¼Œç„¶åå°±èƒ½æ ¹æ®è¿™ä¸ªå¯¹ç»†èƒçš„æ‰¹æ¬¡è¿›è¡Œå¯è§†åŒ–ï¼Œ**ç»“æœæœ€å¥½æ˜¯å„ä¸ªç»†èƒäºšç¾¤çš„ä¸åŒæ‰¹æ¬¡æ··åœ¨ä¸€èµ·**ï¼Œè¿™æ ·æˆ‘ä»¬å°±èƒ½è®¤ä¸ºï¼šé™ç»´æ‰€é‡‡ç”¨çš„â€ç‰¹å¼‚æ€§â€œçš„ç‰¹å¾å¹¶ä¸æ¥è‡ªæ‰¹æ¬¡

```r
plot_cells(cds, color_cells_by="plate", label_cell_groups=FALSE)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-23-014500.png)

è¿™é‡Œçœ‹åˆ°è¿˜ä¸é”™ï¼Œä½†æ˜¯å¦‚æœçœŸçš„æ‹…å¿ƒä¸åŒæ‰¹æ¬¡ä¼šäº§ç”Ÿå…¶ä»–çš„å½±å“ï¼Œå¯ä»¥åœ¨é™ç»´ä¹‹å‰å¯¹æ‰¹æ¬¡è¿›è¡Œæ ¡æ­£ï¼Œä¿è¯åé¢çš„å„ä¸ªç»†èƒç¾¤åªæ¥è‡ªä¸€ä¸ªæ‰¹æ¬¡ï¼Œé‚£ä¹ˆå®ƒä»¬ä¹‹é—´çš„åˆ†ç¾¤å·®å¼‚å°±æ›´å¯èƒ½æ˜¯ç”±äºç”Ÿç‰©å› ç´ å¯¼è‡´

```r
# å‡å…¥è¦å¯¹æ‰¹æ¬¡è¿›è¡Œæ ¡æ­£
cds = preprocess_cds(cds, num_dim = 100, residual_model_formula_str = "~ plate")
# æ ¡æ­£åå†é™ç»´
cds = reduce_dimension(cds)
# å†æ¬¡æ£€æŸ¥æ‰¹æ¬¡æ•ˆåº”
plot_cells(cds, color_cells_by="plate", label_cell_groups=FALSE)
```

**ç»†èƒèšç±» => cluster_cells()**

è¿™æ˜¯ç»†èƒåˆ†ç¾¤è¿‡ç¨‹ä¸­éå¸¸é‡è¦çš„ä¸€æ­¥ã€‚Monocleåˆ©ç”¨[Louvain community detection](https://en.wikipedia.org/wiki/Louvain_Modularity) è¿™ä¸ªéç›‘ç£èšç±»æ–¹æ³•ï¼ˆå®ƒæ˜¯`phenoGraph`ç®—æ³•çš„ä¸€éƒ¨åˆ†ï¼‰ï¼Œå®ƒå’Œæˆ‘ä»¬å¸¸è§çš„Kmeansã€hclustå±‚æ¬¡èšç±»ç­‰è¿˜ä¸å¤ªä¸€æ ·ï¼Œè¿™ä¸ªæ–¹æ³•**æ›´å€¾å‘äºæ‰¾åˆ°ä¸€ä¸ªç½‘ç»œä¸­è”ç³»ç´§å¯†çš„éƒ¨**åˆ†ï¼Œè€Œç»å¸¸å¿½ç•¥èŠ‚ç‚¹çš„ç‰¹æ€§ï¼›è€Œæˆ‘ä»¬**å¸¸è§çš„èšç±»**å‘¢ï¼Œå®ƒ**ä¸€èˆ¬ä¼šå¿½è§†æ•´ä½“ä¸­å„ä¸ªéƒ¨åˆ†çš„è”ç³»ï¼Œ**è€Œé€šè¿‡**è®¡ç®—ä¸¤ä¸ªèŠ‚ç‚¹(ç›®æ ‡)ä¹‹é—´çš„è·ç¦»ï¼ˆå¦‚æ¬§å¼è·ç¦»ã€æ›¼å“ˆé¡¿è·ç¦»ã€ä½™å¼¦ç›¸ä¼¼åº¦ç­‰ï¼‰** æ‰¾åˆ°ç›¸ä¼¼çš„ç‰¹æ€§

è¿™ä¸€æ­¥ä¾èµ–ä¸€ä¸ªPythonæ¨¡å—`louvain`

> **å…³äºPythonæ¨¡å—åœ¨Rä¸­çš„å®‰è£…ï¼š** å¯ä»¥çœ‹æˆ‘å†™è¿‡çš„[Linux/Mac/Windowsçš„Rstudioå®‰è£…Pythonæ¨¡å—æ€»æŠ¥é”™ï¼Œæ€ä¹ˆç ´ï¼Ÿ](https://www.jianshu.com/p/40cc441cd8f5)

åšçš„è¯å¾ˆç®€å•ï¼Œä¸€ä¸ªå‡½æ•°ï¼Œç»“æœä¿å­˜åœ¨äº†`cds@clusters$UMAP$clusters`ä¸­ï¼š

```r
cds = cluster_cells(cds, resolution=c(10^seq(-6,-1)))
plot_cells(cds)
# è¿™ä¸ªå¯è§†åŒ–å‡½æ•°å¦‚æœä¸åŠ ä»»ä½•å‚æ•°ï¼Œå®ƒé»˜è®¤å¯¹ä¸åŒçš„clusterä¸Šè‰²
# é‚£ä¹ˆæœ‰å“ªäº›clusterå‘¢ï¼Ÿç”¨ä¸‹é¢å‡½æ•°æŸ¥çœ‹ï¼š
cds@clusters$UMAP$clusters #æˆ–è€…
clusters(cds) 
#ä¸€å…±æœ‰85ä¸ªcluster
```

![æŒ‰ç…§clusterå¯è§†åŒ–](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-23-030049.png)

ä¸è¿‡è¿™æ ·ç»†åˆ†çœ‹çš„å¤ªæ··æ‚ï¼Œäºæ˜¯ \[Alex Wolf et al]\([https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x)") åˆ©ç”¨ä»–ä»¬å¼€å‘çš„ [PAGA](https://github.com/theislab/paga) ç®—æ³•å°†ä¸€äº›**å°çš„clusterèšåˆ**æˆä¸€ä¸ªå¤§çš„partitionï¼Œç»“æœä¹Ÿæ˜¯ä¿å­˜åœ¨äº†`cluster_cells`çš„è®¡ç®—ç»“æœä¸­`cds@clusters$UMAP$partitions`

```r
plot_cells(cds, color_cells_by="partition", group_cells_by="partition")
# æ£€æŸ¥æœ‰å¤šå°‘partition
cds@clusters$UMAP$partitions
# æ•´åˆæˆäº†36ä¸ªå¤§çš„partition
```

![æŒ‰ç…§partitionè¿›è¡Œå¯è§†åŒ–](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-25-120805.png)

é™¤äº†é»˜è®¤æŒ‰ç…§clusterç¼–å·è¿›è¡Œæ ‡è®°ä¸åŒçš„äºšç¾¤ï¼Œè¿˜å¯ä»¥**æŒ‰ç…§æ¯ä¸ªclusterå¯¹åº”çš„ç»†èƒç±»å‹**è¿›è¡Œå¯è§†åŒ–ï¼š

```r
plot_cells(cds, color_cells_by="cao_cell_type")
```

![æŒ‰ç…§æ¯ä¸ªclusterå¯¹åº”çš„ç»†èƒç±»å‹è¿›è¡Œå¯è§†åŒ–](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-27-020039.png)

ä¸è¿‡è¿™æ ·åˆå¾ˆå¤šé‡å¤çš„ç»†èƒç±»å‹ï¼Œæˆ‘ä»¬å¯ä»¥**å¯¹labelè¿›è¡Œç®€åŒ–ï¼š**

```r
plot_cells(cds, color_cells_by="cao_cell_type", label_groups_by_cluster=FALSE)
```

![å¯¹labelè¿›è¡Œç®€åŒ–](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-27-022233.png)

> **å½’æ ¹ç»“åº•ï¼Œéƒ½æ˜¯å¯¹ç»†èƒçš„ç±»å‹ã€åæ ‡ã€é¢œè‰²æ˜ å°„**ã€‚å°±æ˜¯è¯´ï¼š`cds@colData$cao_cell_type`ä¸­å­˜å‚¨äº†æ¯ä¸ªç»†èƒå¯¹åº”çš„ç»†èƒç±»å‹ï¼›`cds@reducedDims$UMAP`ä¸­å­˜å‚¨äº†æ¯ä¸ªç»†èƒUMAPé™ç»´åçš„åæ ‡ï¼›`cds@clusters$UMAP$partitions`æˆ–`cds@clusters$UMAP$clusters`ä¸­å­˜å‚¨äº†æ¯ä¸ªç»†èƒèšç±»åå¯¹åº”çš„ç»†èƒåˆ†ç¾¤ï¼›æ¥ä¸‹æ¥å°±æ˜¯æ ¹æ®åæ ‡å¯¹æ¯ä¸ªç»†èƒè¿›è¡Œåˆ†ç¾¤ä¸Šè‰²ã€æ ‡è®°ç±»å‹

**æ‰¾markeråŸºå›  => top_markers()**

ç¡®å®šäº†å„ç§äºšç¾¤ä»¥åï¼Œæˆ‘ä»¬å°±æƒ³çŸ¥é“ä»€ä¹ˆå¯¼è‡´äº†å®ƒä»¬çš„å·®å¼‚ï¼Œè¿™å°±æ˜¯å¯»æ‰¾markeråŸºå› çš„è¿‡ç¨‹ï¼Œ**åˆ©ç”¨`top_markers()`å‡½æ•°**

```r
marker_test_res = top_markers(cds, group_cells_by="partition", reference_cells=1000, cores=8)
# è®¾ç½®reference_cellsæ˜¯éšæœºæŒ‘å‡ºæ¥è¿™äº›æ•°é‡çš„ç»†èƒä½œä¸ºå‚ç…§ï¼Œç„¶åè®©top_markerså’Œå‚ç…§é›†ä¸­çš„åŸºå› è¿›è¡Œæ˜¾è‘—æ€§æ£€éªŒï¼›å¦å¤–reference_cellsè¿˜å¯ä»¥æ˜¯æ¥è‡ªcolnames(cds)çš„ç»†èƒå
marker_test_res[1:4,1:4]
```

ç»“æœå¾—åˆ°ä¸€ä¸ªæ•°æ®æ¡†ï¼Œå…¶ä¸­åŒ…å«äº†æ¯ä¸ª`partition` ï¼ˆå› ä¸ºä¸Šé¢æˆ‘ä»¬è®¾ç½®äº†`group_cells_by`ï¼‰çš„ç‰¹å¼‚è¡¨è¾¾åŸºå› ã€‚

ç„¶åæˆ‘ä»¬å¯ä»¥**è‡ªè¡Œè¿‡æ»¤ï¼š**

```r
top_specific_markers = marker_test_res %>%
    filter(fraction_expressing >= 0.10) %>%
    group_by(cell_group) %>%
    top_n(1, pseudo_R2)
# åŸºå› idå»é‡
top_specific_marker_ids = unique(top_specific_markers %>% pull(gene_id))
```

ç„¶åå¯ä»¥**å¯¹æ¯ç»„çš„markeråŸºå› å¯è§†åŒ–=> plot_genes_by_group()ï¼š**

```r
plot_genes_by_group(cds,
                    top_specific_marker_ids,
                    group_cells_by="partition",
                    ordering_type="maximal_on_diag",
                    max.size=3)
```

![å¯¹æ¯ç»„çš„markeråŸºå› å¯è§†åŒ–](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-27-034950.png)

å¦‚æœè¦å¤šçœ‹å‡ ä¸ªmarkerï¼Œåªéœ€è¦ä¿®æ”¹ä¸€ä¸‹`top_n`

```r
top_specific_markers = marker_test_res %>%
    filter(fraction_expressing >= 0.10) %>%
    group_by(cell_group) %>%
    top_n(3, pseudo_R2)

top_specific_marker_ids = unique(top_specific_markers %>% pull(gene_id))

plot_genes_by_group(cds,
                    top_specific_marker_ids,
                    group_cells_by="partition",
                    ordering_type="cluster_row_col",
                    max.size=3)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-27-035540.png)

**å¯¹ç»†èƒç±»å‹è¿›è¡Œæ³¨é‡Š**

å› ä¸ºæˆ‘ä»¬ä¸Šé¢çš„æµ‹è¯•æ•°æ®ä¸­ç»™å‡ºäº†ç»†èƒç±»å‹ï¼Œæ‰€ä»¥å¯ä»¥è¿›è¡Œèšç±»åçš„å¯è§†åŒ–ï¼Œä½†å®é™…ä¸Šï¼Œæˆ‘ä»¬é€šå¸¸åªèƒ½çœ‹åˆ°ä¸åŒçš„clusteræ•°å­—å’Œå®ƒä»¬çš„åˆ†å¸ƒï¼ˆä¹Ÿå°±æ˜¯ç±»ä¼¼äº`as.character(partitions(cds))`çš„ç»“æœï¼‰ï¼Œå› æ­¤**åˆ©ç”¨markeråŸºå› å»é‡æ–°å®šä¹‰ç»†èƒç±»å‹æ˜¯éå¸¸å…³é”®çš„ä¸€æ­¥ã€‚**

```r
# å…ˆå°†partitionsçš„åˆ†ç»„ç”±å› å­å‹è½¬ä¸ºå­—ç¬¦å‹
colData(cds)$assigned_cell_type = as.character(partitions(cds))
# å†å¯¹å­—ç¬¦å‹é‡æ–°å®šä¹‰
colData(cds)$assigned_cell_type = dplyr::recode(colData(cds)$assigned_cell_type,
                                                "1"="Body wall muscle",
                                                "2"="Germline",
                                                "3"="Unclassified neurons",
                                                "4"="Seam cells",
                                                "5"="Coelomocytes",
                                                "6"="Pharyngeal epithelia",
                                                "7"="Vulval precursors",
                                                "8"="Non-seam hypodermis",
                                                "9"="Intestinal/rectal muscle",
                                                "10"="Touch receptor neurons",
                                                "11"="Pharyngeal neurons",
                                                "12"="Am/PH sheath cells",
                                                "13"="NA",
                                                "14"="Unclassified neurons",
                                                "15"="flp-1(+) interneurons",
                                                "16"="Canal associated neurons",
                                                "17"="Pharyngeal gland",
                                                "18"="Other interneurons",
                                                "19"="Ciliated sensory neurons",
                                                "20"="Ciliated sensory neurons",
                                                "21"="Ciliated sensory neurons",
                                                "22"="Ciliated sensory neurons",
                                                "23"="Ciliated sensory neurons",
                                                "24"="Ciliated sensory neurons",
                                                "25"="Oxygen sensory neurons",
                                                "26"="Ciliated sensory neurons",
                                                "27"="Unclassified neurons",
                                                "28"="Pharyngeal gland",
                                                "29"="Ciliated sensory neurons",
                                                "30"="Ciliated sensory neurons",
                                                "31"="Ciliated sensory neurons",
                                                "32"="Ciliated sensory neurons",
                                                "33"="Pharyngeal muscle",
                                                "34"="Failed QC")
plot_cells(cds, group_cells_by="partition", color_cells_by="assigned_cell_type")
# Seuratä¸­ç”¨RenameIdents è¿›è¡Œç»†èƒç±»å‹é‡å®šä¹‰

# å¦å¤–ï¼Œæƒ³ä»ä¸­å–å­é›†ã€è¿‡æ»¤çš„è¯
cds[,colData(cds)$assigned_cell_type != "Failed QC"]
```

![ç»†èƒç±»å‹è¿›è¡Œæ³¨é‡Š](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-27-041026.png)

**åŸºäºGarnettçš„è‡ªåŠ¨åŒ–æ³¨é‡Š**

> ä¸‹é¢è¿™éƒ¨åˆ†å…ˆäº†è§£ä¸‹æ–¹æ³•å°±è¡Œï¼Œæ—¥åæœ‰éœ€è¦å†ç»†å­¦

å‰é¢å¯¹ç»†èƒç±»å‹è¿›è¡Œæ‰‹åŠ¨æ³¨é‡Šæœ‰ç‚¹éº»çƒ¦ï¼Œè€Œä¸”clusterç¼–å·ä¸€æ—¦æ›´æ”¹ï¼Œåˆè¦æ‰‹åŠ¨å»æ³¨é‡Šä¸€éã€‚

åˆ©ç”¨Monocleå›¢é˜Ÿå¼€å‘çš„ [Garnett](https://cole-trapnell-lab.github.io/garnett/) ï¼Œå®ƒå¯ä»¥æ ¹æ®markeråŸºå› å¯¹ç»†èƒè¿›è¡Œåˆ†ç±»

**é¢„å¤„ç†**

```r
# step-1ï¼šé¦–å…ˆæ ¹æ®ä¸Šé¢å¾—åˆ°çš„ç»†èƒç±»å‹assigned_cell_typeï¼Œæ‰¾top_marker
assigned_type_marker_test_res = top_markers(cds,
                                            group_cells_by="assigned_cell_type",
                                            reference_cells=1000,
                                            cores=8)
# step-2ï¼šè¿‡æ»¤ï¼ˆé˜ˆå€¼è‡ªå®šä¹‰ï¼‰
garnett_markers = assigned_type_marker_test_res %>%
  filter(marker_test_q_value < 0.01 & specificity >= 0.5) %>%
  group_by(cell_group) %>%
  top_n(5, marker_score)
# step-3ï¼šå»é‡å¤
garnett_markers = garnett_markers %>% group_by(gene_short_name) %>%
  filter(n() == 1)
# step-4ï¼šç”Ÿæˆmarkeræ–‡ä»¶
generate_garnett_marker_file(garnett_markers, file="./marker_file.txt")
```

![generate_garnett_marker_fileç”Ÿæˆmarkeræ–‡ä»¶](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-27-081221.png)

å¾—åˆ°è¿™ä¸ªæ–‡ä»¶ï¼Œå…¶å®ä¹Ÿåªæ˜¯è‡ªåŠ¨åŒ–å®ç°çš„ç¬¬ä¸€æ­¥ï¼Œæˆ‘ä»¬è‡ªå·±å¯ä»¥æ ¹æ®å·²æœ‰çš„çŸ¥è¯†åœ¨å…¶ä¸­å¢åŠ æˆ–åˆ é™¤ä¸€äº›markerä¿¡æ¯ï¼›å¦å¤–å¦‚æœå‘ç°å¤šä¸ªç»†èƒç±»å‹ä¸­çš„markeråŸºå› æœ‰å¤§é‡é‡åˆï¼Œå°±è¦è€ƒè™‘è¿™å‡ ç§ç»†èƒç±»å‹å¯èƒ½éƒ½æ˜¯æŸä¸€ç§çš„ç»†èƒçš„äºšå‹ã€‚è¿˜æœ‰å¾ˆå¤šå…³äºmarkeræ–‡ä»¶ä¿®æ”¹çš„å¸®åŠ©ä¿¡æ¯ï¼šGarnett [documentation](https://cole-trapnell-lab.github.io/garnett/docs/)

**æ–¹æ³•ä¸€ï¼šè‡ªå·±ä½¿ç”¨Garnett**

```r
# å®‰è£…Garnett
devtools::install_github("cole-trapnell-lab/garnett", ref="monocle3")
library(garnett)
# å®‰è£…ç‰©ç§æ³¨é‡Šåº“(è¿™é‡Œçš„ç‰©ç§æ˜¯C. elegansç§€ä¸½éšæ†çº¿è™«)ï¼Œç›®çš„æ˜¯ä¸ºäº†åšåŸºå› IDè½¬æ¢
BiocManager::install("org.Ce.eg.db")
colData(cds)$garnett_cluster = clusters(cds)
worm_classifier <- train_cell_classifier(cds = cds,
                                         marker_file = "./marker_file.txt",
                                         db=org.Ce.eg.db::org.Ce.eg.db,
                                         cds_gene_id_type = "ENSEMBL",
                                         num_unknown = 50,
                                         marker_file_gene_id_type = "SYMBOL",
                                         cores=8)
# è¿™æ ·è‡ªå·±å°±åšäº†ä¸€ä¸ªåˆ†ç»„æ•°æ®é›†ï¼Œå®˜æ–¹ä¹Ÿå»ºè®®å°†é«˜è´¨é‡çš„è®­ç»ƒé›†æ±‡æ€»åˆ°ï¼šhttps://github.com/cole-trapnell-lab/garnett/issues
# 
cds = classify_cells(cds, worm_classifier,
                           db = org.Ce.eg.db::org.Ce.eg.db,
                           cluster_extend = TRUE,
                           cds_gene_id_type = "ENSEMBL")
plot_cells(cds,
           group_cells_by="partition",
           color_cells_by="cluster_ext_type")
```

**æ–¹æ³•äºŒï¼šä½¿ç”¨åˆ«äººæä¾›çš„åˆ†ç»„æ•°æ®é›†**

ä¾‹å¦‚ä½œè€…æä¾›äº†å·²ç»åšå¥½çš„æ•°æ®é›†ï¼ˆ[https://cole-trapnell-lab.github.io/garnett/classifiers/ceWholeï¼‰ï¼Œæˆ‘ä»¬ç›´æ¥ä¸‹è½½å°±èƒ½ä½¿ç”¨ï¼š](https://cole-trapnell-lab.github.io/garnett/classifiers/ceWhole%EF%BC%89%EF%BC%8C%E6%88%91%E4%BB%AC%E7%9B%B4%E6%8E%A5%E4%B8%8B%E8%BD%BD%E5%B0%B1%E8%83%BD%E4%BD%BF%E7%94%A8%EF%BC%9A)

```r
load(ceWhole)
cds = classify_cells(cds, ceWhole,
                           db = org.Ce.eg.db,
                           cluster_extend = TRUE,
                           cds_gene_id_type = "ENSEMBL")
```

### Monocleç¤ºä¾‹2-æ„å»ºå‘è‚²è½¨è¿¹

**æ„ä¹‰ï¼š** åœ¨å‘è‚²ç”Ÿç‰©å­¦ä¸­ï¼Œç»†èƒæ•´ä¸ªå‘è‚²é˜¶æ®µä¸­å¯¹åˆºæ¿€åšå‡ºçš„ä¸åŒç›¸åº”å¯¼è‡´äº†ç»†èƒåŠŸèƒ½çŠ¶æ€çš„è½¬åŒ–ã€‚ä¸åŒçŠ¶æ€çš„ç»†èƒè¡¨è¾¾ä¸åŒçš„åŸºå› ï¼Œä»è€Œäº§ç”Ÿä¸åŒçš„è›‹ç™½ã€ä»£è°¢ç‰©å‚ä¸ä¸åŒçš„ç”Ÿå‘½è¿‡ç¨‹ã€‚ä¼´éšç€ç»†èƒçŠ¶æ€çš„è½¬åŒ–ï¼Œè½¬å½•ä¿¡æ¯ä¹Ÿåœ¨ä¸æ–­æ›´æ¢ï¼Œæœ‰çš„åŸºå› èµ·åˆæ²‰é»˜åæ¥æ¿€æ´»ã€‚ä½†æ˜¯è¿™ä¸ªäº‹æƒ…å¾ˆéš¾å»æ¢ç´¢ï¼Œå› ä¸ºè¦çº¯åŒ–å‡ºä¸åŒçŠ¶æ€å¹¶ä¸”ç¨³å®šçš„ç»†èƒåŸºæœ¬æ˜¯ä¸ç°å®çš„ï¼Œè€ŒscRNAå¯ä»¥ä¸ç”¨é€šè¿‡å®éªŒæ–¹æ³•çš„çº¯åŒ–ç»†èƒï¼Œå°±èƒ½ç ”ç©¶ç»†èƒç”Ÿå‘½çš„å„ä¸ªçŠ¶æ€ã€‚

**è¿™ä¹Ÿæ˜¯Monocleæœ€å¤§çš„äº®ç‚¹**â€”åˆ©ç”¨ç®—æ³•å»å­¦ä¹ æ¯ä¸ªç»†èƒåŸºå› è¡¨è¾¾é‡çš„å˜åŒ–ï¼Œä»è€Œå¯¹ç»†èƒçŠ¶æ€è¿›è¡Œæ¨æ–­ï¼Œæ¨æ–­å‡ºæ•´ä½“çš„ä¸€ä¸ªè¡¨è¾¾è½¨è¿¹ï¼Œé‚£ä¹ˆå°±å¯ä»¥å°†æ¯ä¸ªç»†èƒæ”¾åœ¨ç‰¹å®šä½ç½®ä¸Šï¼›å¦å¤–æœ‰å¯èƒ½ä¸€ä¸ªè¿‡ç¨‹æœ‰å¤šä¸ªç»“æœï¼Œé‚£ä¹ˆåœ¨è½¨è¿¹å›¾ä¸Šå°±ä¼šæœ‰å¤šä¸ªåˆ†æ”¯ï¼Œä»£è¡¨äº†ç»†èƒåˆ†åŒ–

> **ä¸Šé¢ä½¿ç”¨çš„æ•°æ®é›†åˆ°æ­¤ç»“æŸ**ï¼Œä¸‹é¢å‘è‚²è½¨è¿¹ä¼šç”¨åˆ°ä¸€å¥—æ–°çš„æ•°æ®é›†ï¼ˆæ¥è‡ª[Packer & Zhu et al](http://dx.doi.org/10.1101/565549). çš„çº¿è™«æ•´ä¸ªèƒšèƒå‘è‚²çš„æ—¶é—´çº¿æ•°æ®é›†ï¼Œè¿™é‡Œä¸»è¦èŠ‚é€‰äº†ç¥ç»å…ƒçš„æ•°æ®ï¼‰ **éœ€è¦æ³¨æ„çš„æ˜¯ï¼Œä¸‹é¢åšå‡ºçš„å›¾å’Œå®˜ç½‘ç»™å‡ºçš„å›¾æ˜¯ä¸åŒçš„**

**è½½å…¥æ•°æ®ï¼Œæ„å»ºå¯¹è±¡ => new_cell_data_set()**

```r
expression_matrix = readRDS(url("http://staff.washington.edu/hpliner/data/packer_embryo_expression.rds"))
cell_metadata = readRDS(url("http://staff.washington.edu/hpliner/data/packer_embryo_colData.rds"))
gene_annotation = readRDS(url("http://staff.washington.edu/hpliner/data/packer_embryo_rowData.rds"))

# æ–‡ä»¶å¤§å°(åŒæ ·å¯ä»¥å…¬ä¼—å·åå°å›å¤"monocle-2"è·å–)
#|-- [497K]  packer_embryo_colData.rds
#|-- [10.0M]  packer_embryo_expression.rds
#`-- [226K]  packer_embryo_rowData.rds

cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
```

**é¢„å¤„ç† => preprocess_cds()**

å’Œç¤ºä¾‹ä¸€çš„æ“ä½œå·®ä¸å¤šï¼Œåªæ˜¯è¿™é‡Œæ ¹æ®åŸä½œè€…ï¼ˆPacker & Zhu et al ï¼‰çš„æ–¹æ³•å¤„ç†äº†ä¸€ä¸‹æ‰¹æ¬¡æ•ˆåº”

```r
cds <- preprocess_cds(cds, num_dim = 100, residual_model_formula_str = "~ bg.300.loading + bg.400.loading + bg.500.1.loading + bg.500.2.loading + bg.r17.loading + bg.b01.loading + bg.b02.loading")
```

**é™ç»´å¯è§†åŒ– => reduce_dimension()**

å¼ºçƒˆå»ºè®®ç”¨UMAP

```r
cds <- reduce_dimension(cds)
plot_cells(cds, label_groups_by_cluster=FALSE,  color_cells_by = "cell.type")
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-27-101608.png)

çœ‹ä¸åŒåŸºå› åœ¨ä¸åŒåˆ†æ”¯çš„è¡¨è¾¾é‡:

```r
ciliated_genes = c("che-1",
                   "hlh-17",
                   "nhr-6",
                   "dmd-6",
                   "ceh-36",
                   "ham-1")

plot_cells(cds,
           genes=ciliated_genes,
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-013917.png)

**ç»†èƒèšç±» => cluster_cells()**

Monocleä¸è®¤ä¸ºä¸€ç»„æ•°æ®ä¸­çš„æ‰€æœ‰ç»†èƒéƒ½æ¥è‡ªåŒä¸€ä¸ªâ€ç¥–å…ˆâ€œï¼Œå¾ˆå¤šå®éªŒä¸­ï¼Œå®ƒä»¬ä¼šæœ‰å¤šä¸ªå‘è‚²è½¨è¿¹ã€‚Monocleä¼šé€šè¿‡èšç±»æ¥åˆ¤æ–­ç»†èƒæ˜¯å¦åº”è¯¥å½’å±åŒä¸€ä¸ªå‘è‚²è½¨è¿¹ã€‚ä¹‹å‰ä»‹ç»çš„`cluster_cells()`ä¸­ï¼Œç»†èƒå¯ä»¥æŒ‰clusterç»†åˆ†ï¼Œè¿˜å¯ä»¥æŒ‰partitionå½’ä¸ºå¤§ç±»ã€‚

```r
# è¿™é‡Œå°±ç”¨partitionèšç±»
cds <- cluster_cells(cds)
plot_cells(cds, color_cells_by = "partition")
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-014614.png)

**åˆ©ç”¨learn graphåœ¨æ¯ä¸ªpartitionä¸­å¯»æ‰¾ä¸»è·¯å¾„ => learn_graph()**

> ç†è®ºä¸Šï¼Œä»»ä½•çš„UMAPæˆ–è€…tSNEé™ç»´ç»“æœéƒ½èƒ½æ‰¾åˆ°è¿™æ ·çš„ä¸»è·¯å¾„ï¼Œé€‰ç”¨UMAPæ˜¯è€ƒè™‘äº†å®ƒèƒ½æ›´å¤šå±•ç¤ºç»†èƒä¹‹é—´çš„å…³è”ï¼Œå¦å¤–æœ€é‡è¦æ˜¯å¯¹æ•°æ®çš„äº†è§£

```r
cds <- learn_graph(cds)
plot_cells(cds,
           color_cells_by = "cell.type",
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-021732.png)

**æœ‰äº†åˆæ­¥çš„è½¨è¿¹çº¿ï¼Œæ¥ä¸‹æ¥å°±æ˜¯å¯¹ç»†èƒå‡ºç°çš„å…ˆåè¿›è¡Œæ’åº => order_cells()**

> ä¹Ÿå°±æ˜¯pseudotimeï¼ˆæ‹Ÿæ—¶åºåˆ†æï¼‰ï¼Œæ¨æ–­å“ªä¸ªç»†èƒå…ˆå‡ºæ¥ï¼Œå“ªä¸ªç»†èƒåå‡ºæ¥ã€‚å®ƒä¼šå¯¹æ¯ä¸ªç»†èƒåœ¨ç‰¹å®šçš„è¿‡ç¨‹ï¼ˆæ¯”å¦‚åˆ†åŒ–è¿‡ç¨‹ï¼‰ä¸­çš„å‚ä¸åº¦è¿›è¡Œè¯„ä¼°

å¾ˆå¤šç”Ÿå‘½æ´»åŠ¨ä¸­ï¼Œç»†èƒå¹¶éåŒæ­¥å‘è‚²çš„ï¼Œä¾‹å¦‚åœ¨ç»†èƒåˆ†åŒ–é˜¶æ®µï¼Œå®éªŒæ•è·åˆ°çš„ç»†èƒå¯èƒ½ä¼šåˆ†å¸ƒåˆ°å„ä¸ªå‘è‚²è¿‡ç¨‹ã€‚å› æ­¤åšå•ç»†èƒæµ‹åºæ—¶å¾—åˆ°çš„é‚£ä¸€ç¾¤ç»†èƒï¼Œå…¶ä¸­å¯èƒ½åŒ…æ‹¬å·²ç»å®ŒæˆæŸä¸ªç”Ÿå‘½è¿‡ç¨‹çš„ç»†èƒï¼Œè¿˜å¯èƒ½åŒ…æ‹¬æ²¡æœ‰å¼€å§‹è¿™ä¸ªè¿‡ç¨‹çš„ç»†èƒã€‚è¿™ç§éåŒæ­¥çš„ç‰¹æ€§ä¼šé˜»ç¢æˆ‘ä»¬ç†è§£ç»†èƒè¿‡æ¸¡çŠ¶æ€ç©¶ç«Ÿå‘ç”Ÿäº†æ€æ ·çš„è°ƒæ§ã€‚

Monocleçš„è¿™ä¸ªç®—æ³•å°†åŸºå› è¡¨è¾¾é‡çš„å˜åŒ–å®šä¹‰ä¸ºå‘è‚²è½¨è¿¹ä¸Š**ç”Ÿå‘½è¿‡ç¨‹çš„å˜åŒ–**ï¼ˆä½“ç°ä¸ºä¸åŒçš„ç»†èƒç±»å‹ï¼‰**ï¼Œè€Œä¸æ˜¯çœŸæ­£çš„æ—¶é—´å˜åŒ–**ï¼ˆå› æ­¤æˆ‘ä»¬å«å®ƒâ€æ‹Ÿæ—¶åºâ€œï¼Œè€Œä¸æ˜¯â€çœŸæ—¶é—´åˆ†æâ€œï¼‰ã€‚

ä¹‹å‰å¾—åˆ°çš„è½¨è¿¹æ€»é•¿åº¦å°±æ˜¯ç»†èƒä»ä¸€ä¸ªé˜¶æ®µçš„èµ·å§‹åˆ°ç»ˆæ­¢æ‰€äº§ç”Ÿçš„è½¬å½•æ°´å¹³å˜åŒ–æ€»é‡ï¼›æ‹Ÿæ—¶åºåˆ†æä¼šæ²¿ç€æœ€çŸ­çš„è·¯å¾„ï¼Œè®¡ç®—æ¯ä¸ªç»†èƒå’Œè½¨è¿¹èµ·å§‹ä½ç‚¹çš„è·ç¦»

```r
# å…ˆå°†æ¯ä¸ªç»†èƒæ ¹æ®â€èƒšèƒå‘è‚²æ—¶é—´åŒºé—´â€œè¿›è¡Œä¸Šè‰²ï¼Œç„¶åæ ¹æ®èƒšèƒå‘è‚²è¿‡ç¨‹å‰åç»˜åˆ¶èŠ‚ç‚¹
# åœ¨cds@colDataä¸­æä¾›äº†ç»†èƒå¯¹åº”çš„ç±»å‹ä»¥åŠèƒšèƒå‘è‚²æ—¶é—´ï¼Œè¿™ä¸ªæ‰æ˜¯çœŸæ­£çš„å…³é”®ä¿¡æ¯ã€‚æœ‰äº†è¿™ä¸ªä¿¡æ¯æ‰èƒ½ä½œå›¾ã€‚å·¥å…·ä¸æ˜¯ä¸‡èƒ½çš„ï¼Œå®ƒä¸å¯èƒ½å¸®åŠ©æˆ‘ä»¬å»å®Œæˆç”Ÿç‰©å­¦ä¸­çš„æ¨æ–­ï¼Œä¸€å®šæ˜¯æˆ‘ä»¬è‡ªå·±å…ˆå®šä¹‰å¥½ï¼Œå†äº¤ç»™å®ƒè¿›è¡Œå¯è§†åŒ–è€Œå·²
plot_cells(cds,
           color_cells_by = "embryo.time.bin",
           label_cell_groups=FALSE,
           label_leaves=TRUE,
           label_branch_points=TRUE,
           graph_label_size=1.5)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-025246.png)

* å›¾ä¸­çš„**é»‘çº¿**å°±æ˜¯æ•´ä¸ªæ¶æ„ï¼ˆæ³¨æ„åˆ°æ•´ä¸ªå›¾å¹¶éå®Œå…¨è¿æ¥çš„ï¼Œæ¯•ç«Ÿæ˜¯æŒ‰ç…§partitionèšç±»ï¼Œæ¯ä¸ªpartitionä¸­çš„ç»†èƒå·®å¼‚æœ‰ç‚¹å¤§ï¼‰ï¼›
* **æµ…ç°è‰²çš„åœ†åœˆ**è¡¨ç¤ºâ€å¶ç‰‡â€œè¡¨ç¤ºå‘è‚²è½¨è¿¹ä¸­çš„ä¸åŒç»“å±€ï¼ˆå¯ä»¥è®¤ä¸ºæ‹Ÿæ—¶åºåˆ†æçš„å›¾æ˜¯ä¸€ä¸ªâ€æ ¹-èŒ-å¶â€œç»“æ„ï¼‰ï¼Œç”¨å‚æ•°`label_leaves`æ§åˆ¶ï¼›
* **é»‘è‰²çš„åœ†åœˆ**è¡¨ç¤ºåˆ†æ”¯èŠ‚ç‚¹ï¼Œé¢„ç¤ºç€å…¶ä¸­çš„ç»†èƒä¼šæœ‰ä¸åŒå‘å±•æ–¹å‘ï¼Œç”¨å‚æ•°`label_branch_points`æ§åˆ¶ï¼›
* åœˆä¸­æ•°å­—å¤§å°è¡¨ç¤ºå‡ºç°æ—¶é—´çš„å…ˆå

ä»ä¸Šé¢çš„å›¾ä¸­æˆ‘ä»¬å°±èƒ½çŸ¥é“å‡ºç°æ—¶é—´æ¯”è¾ƒæ—©çš„ç»†èƒä½ç½®ï¼Œæˆ‘ä»¬éœ€è¦å¯¹å…¨éƒ¨çš„ç»†èƒéƒ½æ’ä¸ªå…ˆåé¡ºåºï¼Œå› æ­¤è¦ç”¨åˆ°`order_cells()` ï¼Œä¸è¿‡è¿™ä¸ªå‡½æ•°**éœ€è¦ä¸€ä¸ªâ€æ ¹èŠ‚ç‚¹â€œä½ç½®**ï¼Œä¸€èˆ¬æ˜¯ä¸€ä¸ªpartitionåˆ†é…ä¸€ä¸ªæ ¹èŠ‚ç‚¹ã€‚

ä¸è¿‡ï¼Œ**æ€ä¹ˆç®€å•é«˜æ•ˆæŒ‘é€‰æ ¹èŠ‚ç‚¹å‘¢ï¼Ÿ**=> get_earliest_principal_node()

é¦–å…ˆæ ¹æ®è½¨è¿¹å›¾ä¸­çš„èŠ‚ç‚¹å°†ä¸å®ƒä»¬æœ€ç›¸è¿‘çš„ç»†èƒåˆ†æˆç»„ï¼Œç„¶åè®¡ç®—æ¥è‡ªæœ€æ—©æ—¶é—´ç‚¹çš„ç»†èƒæ‰€å ç»„åˆ†ï¼Œæœ€åæŒ‘å‡ºåŒ…å«æ—©æœŸç»†èƒæ•°é‡æœ€å¤šçš„èŠ‚ç‚¹ï¼Œè®¤ä¸ºå®ƒæ˜¯å°±æ˜¯æ ¹èŠ‚ç‚¹

```r
# å®˜æ–¹ç»™å‡ºäº†ä¸€ä¸ªå‡½æ•°ï¼Œè¿™é‡Œå®šä¹‰äº†ä¸€ä¸ªtime_binï¼Œé€‰æ‹©äº†æœ€æ—©çš„æ—¶é—´ç‚¹åŒºé—´ã€‚
get_earliest_principal_node <- function(cds, time_bin="130-170"){
  # é¦–å…ˆæ‰¾åˆ°å‡ºç°åœ¨æœ€æ—©æ—¶é—´åŒºé—´çš„ç»†èƒID
  cell_ids <- which(colData(cds)[, "embryo.time.bin"] == time_bin)

  closest_vertex <-
    cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes <-
    igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names
      (which.max(table(closest_vertex[cell_ids,]))))]

  root_pr_nodes
}
cds = order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))
```

ç„¶åè¿›è¡Œå¯è§†åŒ–ï¼š

```r
plot_cells(cds,
           color_cells_by = "pseudotime",
           label_cell_groups=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           graph_label_size=1.5)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-063403.png)

**åˆ©ç”¨3Dçš„å‘è‚²è½¨è¿¹å¯¹ä¸Šè¿°å†…å®¹åšä¸ªæ¦‚è¿°**

3Dè½¨è¿¹å®é™…ä¸Šå°±æ˜¯é™ç»´æ—¶é€‰å‰3ä¸ªä¸»æˆåˆ† => `max_components = 3`ï¼Œåç»­éƒ½å’Œ2Dä¿æŒç±»ä¼¼

> éœ€è¦æ³¨æ„çš„æ˜¯ï¼Œè¿™ä¸ª3Då›¾éœ€è¦ç­‰å¾…ä¸€æ®µæ—¶é—´åŠ è½½ï¼ŒåŠ è½½å‡ºæ¥çš„å›¾å¯ä»¥ç”¨é¼ æ ‡æ‹–åŠ¨

```r
# 3D trajectoriesï¼ˆ4æ­¥èµ°ï¼‰
cds_3d = reduce_dimension(cds, max_components = 3)
cds_3d = cluster_cells(cds_3d)
cds_3d = learn_graph(cds_3d)
cds_3d = order_cells(cds_3d, root_pr_nodes=get_earliest_principal_node(cds))

cds_3d_plot_obj = plot_cells_3d(cds_3d, color_cells_by="partition")
cds_3d_plot_obj
```

![3Då‘è‚²è½¨è¿¹](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-063935.png)

### Monocleå·®å¼‚åˆ†æ

> **ç‰ˆæœ¬3å’Œç‰ˆæœ¬2çš„å·®å¼‚åˆ†æå¯ä»¥è¯´æ˜¯å®Œå…¨ä¸åŒ**ï¼Œç‰ˆæœ¬3å–ä»£äº†2ä¸­çš„`differentialGeneTest()` and `BEAM()`

Monocle3æä¾›äº†ä¸åŒç»†èƒç±»å‹ä¹‹é—´å¯»æ‰¾å·®å¼‚åŸºå› çš„æ–¹æ³•ï¼Œä¸»è¦æœ‰ä¸¤ç§ï¼š

* Regression analysisï¼šåˆ©ç”¨`fit_models()`ï¼Œç”¨æ¥è¯„ä»·åŸºå› è¡¨è¾¾æ˜¯å¦ä¼šå—åˆ°è¯¸å¦‚æ—¶é—´ã€å¤„ç†ç­‰çš„å½±å“
* Graph-autocorrelation analysisï¼šåˆ©ç”¨`graph_test()`ï¼Œç”¨æ¥å¯»æ‰¾ä¸€æ¡è½¨è¿¹æˆ–ä¸åŒclusterä¸­åŸºå› çš„å·®å¼‚

**æ–¹æ³•ä¸€ï¼šRegression analysis => fit_models()**

> è¿™ä¸€éƒ¨åˆ†å°†ä¼šæ ¹æ®å‡ ç§ä¸åŒçš„æ ‡å‡†è¿›è¡Œå·®å¼‚åˆ†æï¼Œå¹¶ä¸”ä¸åŒå¤æ‚ç¨‹åº¦çš„åˆ†æç”¨æ—¶ä¹Ÿæœ‰è¾ƒå¤§å·®åˆ«ï¼ˆå‡ åˆ†é’Ÿç”šè‡³å‡ å°æ—¶ï¼‰ã€‚æ•™ç¨‹åªä¼šé’ˆå¯¹ä¸€å°éƒ¨åˆ†åŸºå› è¿›è¡Œæœ€ç®€å•çš„åˆ†æï¼Œä½†å®é™…ä¸Šå®ƒå¯ä»¥å¤„ç†ä¸ŠåƒåŸºå› 

æµ‹è¯•å°åŸºå› é›†ä¸ºï¼š

```r
ciliated_genes = c("che-1",
                   "hlh-17",
                   "nhr-6",
                   "dmd-6",
                   "ceh-36",
                   "ham-1")
cds_subset = cds[rowData(cds)$gene_short_name %in% ciliated_genes,]
```

æ¥ä¸‹æ¥ï¼ŒMonocleä¼šå¯¹æ¯ä¸ªåŸºå› æ‹Ÿåˆä¸€ä¸ªå›å½’æ¨¡å‹ï¼Œå…¶ä¸­ä¼šåŠ å…¥å®éªŒä¸­çš„ä¸€äº›å› ç´ ï¼ˆæ¯”å¦‚æ—¶é—´ã€å¤„ç†ç­‰ï¼‰ã€‚ä¾‹å¦‚ï¼Œåœ¨èƒšèƒç›¸å…³çš„å•ç»†èƒæ•°æ®ä¸­ï¼Œç»†èƒä¼šåœ¨ä¸åŒçš„æ—¶é—´ç‚¹è¿›è¡Œæ”¶é›†ï¼Œè¿™é‡Œå°±å¯ä»¥æ£€æµ‹æ˜¯å¦æœ‰åŸºå› åœ¨è¿™æ®µæ—¶é—´å†…çš„è¡¨è¾¾é‡å‘ç”Ÿäº†å˜åŒ–ï¼Œåˆ©ç”¨`fit_models`

```r
gene_fits = fit_models(cds_subset, model_formula_str = "~embryo.time")
# å…¶ä¸­model_formula_strå°±æ˜¯è¦æ¯”è¾ƒçš„åˆ†ç»„å¯¹è±¡ï¼Œå¦‚æœç›¸è·å¾—ä¸åŒçš„clusteræˆ–è€…partitionçš„å·®å¼‚åŸºå› ï¼Œå°±ç”¨model_formula_str = "~cluster"æˆ–è€…model_formula_str = "~partition"ï¼›å¦å¤–è¿˜æ”¯æŒæ·»åŠ å¤šä¸ªå˜é‡ï¼Œæ¯”å¦‚è€ƒè™‘åˆ°æ‰¹æ¬¡æ•ˆåº” model_formula_str = "~embryo.time + batch"
```

ç„¶åçœ‹çœ‹å“ªäº›åŸºå› æ˜¯ä¸æ—¶é—´å› ç´ ç›¸å…³çš„ï¼š

```r
fit_coefs = coefficient_table(gene_fits)
# æŒ‘å‡ºæ—¶é—´ç›¸å…³çš„ç»„åˆ†
emb_time_terms = fit_coefs %>% filter(term == "embryo.time")
# coefficient_table()é»˜è®¤ä½¿ç”¨ Benjamini and Hochbergï¼ˆBHï¼‰æ–¹æ³•è¿›è¡Œäº†på€¼çš„æ ¡æ­£ï¼Œå¾—åˆ°äº†qå€¼
emb_time_terms %>% filter (q_value < 0.05) %>%
  select(gene_short_name, term, q_value, estimate)

# ç»“æœæ•°æ®é›†ä¸­çš„6ä¸ªåŸºå› æœ‰5ä¸ªæ˜¯æ—¶é—´ç›¸å…³çš„å·®å¼‚åŸºå› 
```

é™¤æ­¤ä»¥å¤–ï¼Œè¿˜èƒ½**ç”»å°æç´å›¾ï¼š => plot_genes_violin()**

```r
plot_genes_violin(cds_subset, group_cells_by="embryo.time.bin", ncol=2) +
    theme(axis.text.x=element_text(angle=45, hjust=1))
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-095900.png)

**å¦‚æœè¦è€ƒè™‘æ‰¹æ¬¡æ•ˆåº”**

```r
gene_fits = fit_models(cds_subset, model_formula_str = "~embryo.time + batch")
fit_coefs = coefficient_table(gene_fits)
fit_coefs %>% filter(term != "(Intercept)") %>%
  select(gene_short_name, term, q_value, estimate)
```

**æ¨¡å‹åšå‡ºæ¥äº†ï¼Œæ•ˆæœåˆ°åº•å¦‚ä½•ï¼Ÿ => evaluate_fits()**

å¯¹æ¨¡å‹è¿›è¡Œè¯„ä»·ï¼Œä½¿ç”¨`evaluate_fits()`

```r
gene_fits = fit_models(cds_subset, model_formula_str = "~embryo.time")
evaluate_fits(gene_fits)
```

é‚£ä¹ˆæ¨¡å‹ä¸­åˆ°åº•è¦ä¸è¦è€ƒè™‘æ‰¹æ¬¡å‘¢ï¼Ÿä½¿ç”¨`compare_models()`æ¯”è¾ƒåˆ¤æ–­ã€‚ç»“æœä¼šè¿”å›ä¸€ä¸ªlikelihood ratio testç»“æœï¼Œ

```r
time_batch_models = fit_models(cds_subset,
                               model_formula_str = "~embryo.time + batch",
                               expression_family="negbinomial")
time_models = fit_models(cds_subset,
                        model_formula_str = "~embryo.time",
                        expression_family="negbinomial")
compare_models(time_batch_models, time_models) %>% select(gene_short_name, q_value)
```

å…¶ä¸­ç¬¬ä¸€ä¸ª`time_batch_models`å«åš`full model`ï¼Œè¿™ä¸ªæ¨¡å‹çŸ¥é“ç»†èƒçš„æ”¶é›†æ—¶é—´å’Œæ‰¹æ¬¡ï¼Œæ¥é¢„æµ‹å…¶ä¸­æ¯ä¸ªåŸºå› çš„è¡¨è¾¾ï¼›ç¬¬äºŒä¸ª`time_models`å«åš`reduced model`ï¼Œè¿™ä¸ªæ¨¡å‹åªçŸ¥é“ç»†èƒçš„æ”¶é›†æ—¶é—´ã€‚å› ä¸ºç¬¬ä¸€ä¸ªæ¨¡å‹ä¸­ä¿¡æ¯æ›´ä¸°å¯Œï¼Œæ‰€ä»¥å®ƒå¯¹æ¯ä¸ªç»†èƒä¸­åŸºå› è¡¨è¾¾é¢„æµ‹ä¹Ÿæ›´å‡†ã€‚è€Œ**Monocleè¦å›ç­”çš„é—®é¢˜æ˜¯ï¼šç¬¬ä¸€ä¸ªæ¨¡å‹ç›¸æ¯”äºç¬¬äºŒä¸ªï¼Œç©¶ç«Ÿåšäº†å¤šå°‘æå‡ï¼Ÿ** **å¦‚æœæå‡çš„éƒ¨åˆ†å¤§å¤šæ¥è‡ªç¬¬ä¸€ä¸ªæ¨¡å‹çš„æ‰¹æ¬¡æ•ˆåº”ï¼Œé‚£ä¹ˆå¾—åˆ°çš„likelihood ratio testç»“æœå°±è¶Šæ˜¾è‘—**

```r
gene_short_name  q_value
  <chr>              <dbl>
1 ham-1           3.74e-20
2 dmd-6           7.11e-37
3 hlh-17          1.04e- 5
4 nhr-6           5.95e- 3
5 ceh-36          6.23e-10
6 che-1           5.06e-6
```

ç»“æœçœ‹åˆ°ï¼Œ**æ‰€æœ‰åŸºå› çš„likelihood ratio testsç»“æœéƒ½æ˜¯ææ˜¾è‘—ï¼Œè¯´æ˜äº†æ•°æ®ä¸­çš„ç¡®å­˜åœ¨æ˜¾è‘—çš„æ‰¹æ¬¡æ•ˆåº”**ï¼Œä¹ŸéªŒè¯äº†æ·»åŠ æ‰¹æ¬¡ä¿¡æ¯çš„å¯é æ€§ã€‚

> ä¸Šé¢ğŸ‘†æ‰€ä½¿ç”¨çš„**çº¿æ€§æ¨¡å‹æ–¹æ³•è¦æŒ‡å®šä¸€ä¸ªåŸºå› è¡¨è¾¾é‡çš„åˆ†å¸ƒ**ï¼Œå¤§å¤šæ•°ç ”ç©¶ä½¿ç”¨è´ŸäºŒé¡¹åˆ†å¸ƒï¼Œè¿™ç§åˆ†å¸ƒå¾€å¾€æ˜¯ç¬¦åˆæµ‹åºreadsæˆ–UMI countå€¼çš„ï¼Œè¿™ä¸ªæ–¹æ³•ä¹Ÿè¢«åº”ç”¨åœ¨RNA-seqåˆ†æï¼ˆå¦‚DESeq2ï¼‰ä¸­

`fit_models()`å‡½æ•°æ”¯æŒå¤šç§åˆ†å¸ƒï¼Œé»˜è®¤æ˜¯ï¼š`quasipoisson` ï¼Œå®ƒå’Œè´ŸäºŒé¡¹åˆ†å¸ƒç›¸ä¼¼ï¼Œåªä¸è¿‡æ¯”è´ŸäºŒé¡¹åˆ†å¸ƒå‡†ç¡®æ€§ä½ä¸€ç‚¹ã€é€Ÿåº¦ä¼šæ›´å¿«ï¼Œå¯¹äºç»†èƒæ•°é‡å¤šæ›´é€‚ç”¨ã€‚

å…³äºå…¶ä»–çš„åˆ†å¸ƒï¼ˆä½¿ç”¨`expression_family`å‚æ•°è®¾ç½®ï¼‰ï¼š

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-102243.png)

**æ–¹æ³•äºŒï¼šGraph-autocorrelation analysis => graph_test()**

> æ³¨æ„ï¼šä¸‹é¢çš„ä»£ç å’Œå›¾ç‰‡çœ‹çœ‹å°±å¥½ï¼ˆå› ä¸ºä¸èƒ½è¿è¡Œï¼‰

```r
# é€‰ä¸€éƒ¨åˆ†ç¥ç»å…ƒç»†èƒ
neurons_cds = cds[,colData(cds)$assigned_cell_type == "Neurons"]
plot_cells(neurons_cds, color_cells_by="partition")
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-133349.png)

ä»å›¾ä¸­å¯ä»¥çœ‹åˆ°ï¼Œå­˜åœ¨ä¸åŒç±»å‹çš„ç¥ç»å…ƒç»†èƒï¼Œå®ƒä»¬ä¹Ÿè®¸æ¥è‡ªä¸åŒäºšç¾¤ã€‚ä¸ºäº†æ¢ç´¢è¿™äº›ä¸åŒç±»å‹çš„ç»†èƒä¹‹é—´æ˜¯å“ªäº›åŸºå› å·®å¼‚è¡¨è¾¾ï¼Œä¸€ç§æ–¹æ³•å°±æ˜¯ä¸Šé¢çš„çº¿æ€§å›å½’ã€‚ä¸è¿‡å¯¹äºUMAPæˆ–tSNEçš„å›¾ï¼Œ**`graph_test()`è¿™ä¸ªå‡½æ•°å¯ä»¥è¿›è¡Œä¸€ä¸ªå«åš\[Moran's I]\(**[https://en.wikipedia.org/wiki/Moran's_I](https://en.wikipedia.org/wiki/Moran's_I)**) çš„spatial autocorrelationåˆ†ææ–¹æ³•**

> `Moran's I`è¿™ä¸ªæŒ‡æ ‡å«ï¼šè«å…°æŒ‡æ•°ï¼Œæ¾³å¤§åˆ©äºšç»Ÿè®¡å­¦å®¶è«å…°äº1950å¹´æå‡ºã€‚å®ƒæ˜¯ç”¨æ¥**åº¦é‡ç©ºé—´ç›¸å…³æ€§**çš„ä¸€ä¸ªæŒ‡æ ‡ï¼Œæ•°æ®ç»è¿‡æ–¹å·®å½’ä¸€åŒ–ä¹‹åï¼Œä¼šè½åœ¨-1.0å’Œ1.0ä¹‹é—´ã€‚è¿™ä¸ªå€¼å¤§äº0è¡¨ç¤ºç©ºé—´æ­£ç›¸å…³ï¼Œå€¼è¶Šå¤§ç©ºé—´ç›¸å…³æ€§è¶Šå¼ºï¼›ç­‰äº0æ—¶ç©ºé—´æ˜¯éšæœºæ€§åˆ†å¸ƒï¼›å°äº0æ—¶è¡¨ç¤ºç©ºé—´è´Ÿç›¸å…³ï¼Œå€¼è¶Šå°ç©ºé—´å·®å¼‚è¶Šå¤§

```r
pr_graph_test_res = graph_test(neurons_cds, neighbor_graph="knn", cores=8)
pr_deg_ids = row.names(subset(pr_graph_test_res, q_value < 0.05))
```

`pr_graph_test_res`ç»“æœæ˜¯ä¸€ä¸ªæ•°æ®æ¡†ï¼Œå­˜å‚¨äº†cdså¯¹è±¡ä¸­æ¯ä¸ªåŸºå› çš„`Moran's I`æ£€éªŒç»“æœã€‚æ­£å€¼è¡¨ç¤ºåŸºå› åˆ†å¸ƒåœ¨UMAPç©ºé—´çš„ä¸€å°å—â€ç„¦ç‚¹â€œåŒºåŸŸï¼ˆä¾‹å¦‚ä¸€ä¸ªæˆ–å‡ ä¸ªclusterï¼‰ã€‚

![pr_graph_test_resç»“æœ](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-124940.png)

ç°åœ¨æœ‰äº†å’Œâ€ç„¦ç‚¹â€œåŒºåŸŸå’ŒåŸºå› ï¼Œé‚£ä¹ˆæ€ä¹ˆæŠŠäºŒè€…ç»“åˆèµ·æ¥ï¼Ÿ

**å°†å…±åŒä½œç”¨çš„åŸºå› åˆå¹¶æˆæ¨¡å— => find_gene_modules()**

> æ ¹æ®è¿™äº›å…±åŒä½œç”¨åŸºå› çš„è¡¨è¾¾é‡å¯¹æ‰€æœ‰ç»†èƒè¿›è¡ŒUMAPï¼Œç„¶ååˆ©ç”¨Louvain community analysiså°†å®ƒä»¬èšé›†æˆå°æ¨¡å—ï¼Œç„¶åå°†å°æ¨¡å—å’Œå¤§clusterç”šè‡³æ›´å¤§çš„partitionè”ç³»èµ·æ¥

```r
gene_module_df = find_gene_modules(neurons_cds[pr_deg_ids,], resolution=1e-2)
```

å¾—åˆ°çš„ç»“æœ`gene_module_df`ä¸­æœ‰ä¸€è¡Œè®°å½•äº†æ¯ä¸ªåŸºå› å’Œmoduleçš„å¯¹åº”å…³ç³»

è¦çœ‹å“ªä¸ªmoduleå’Œå“ªä¸ªcluster/partitionç›¸å…³ï¼Œæœ‰**ä¸¤ç§å¯è§†åŒ–æ–¹æ³•ï¼š**

*   ç¬¬ä¸€ç§ï¼š

    ```r
    cell_group_df = tibble::tibble(cell=row.names(colData(neurons_cds)), cell_group=partitions(cds)[colnames(neurons_cds)])
    agg_mat = aggregate_gene_expression(neurons_cds, gene_module_df, cell_group_df)
    row.names(agg_mat) = stringr::str_c("Module ", row.names(agg_mat))
    colnames(agg_mat) = stringr::str_c("Partition ", colnames(agg_mat))

    pheatmap::pheatmap(agg_mat, cluster_rows=TRUE, cluster_cols=TRUE,
                       scale="column", clustering_method="ward.D2",
                       fontsize=6)
    ```

    ç»“æœå¤§ä½“æ˜¯è¿™æ ·ï¼š

    ![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-140341.png)
*   ç¬¬äºŒç§ï¼šé’ˆå¯¹å°‘æ•°é‡çš„moduleå¯ä»¥çœ‹å¾—æ¯”è¾ƒæ¸…æ¥šï¼Œè¿™é‡Œé€‰äº†4ä¸ª

    ```r
    plot_cells(neurons_cds,
               genes=gene_module_df %>% filter(module %in% c(16,38,33,42)),
               group_cells_by="partition",
               color_cells_by="partition",
               show_trajectory_graph=FALSE)
    ```

    ![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-140805.png)

**æ‰¾åˆ°å½±å“å‘è‚²è½¨è¿¹çš„åŸºå› **

> è¿™é‡Œå¾ˆé‡è¦ï¼Œä½†å­¦å®Œåå¹¶æ²¡æœ‰æç„¶å¤§æ˜ç™½çš„æ„Ÿè§‰ï¼Œæ„Ÿè§‰è·Ÿä¸ä¸Šä½œè€…çš„æ€ç»´äº†ã€‚è¿˜è¦ç»§ç»­æ‘¸ç´¢è¿™éƒ¨åˆ†

```r
ciliated_cds_pr_test_res = graph_test(cds, neighbor_graph="principal_graph", cores=4)
# ä½¿ç”¨neighbor_graph="principal_graph"æ¥æ£€éªŒè½¨è¿¹ç›¸é‚»çš„ç»†èƒçš„è¡¨è¾¾æ˜¯å¦ç›¸å…³
pr_deg_ids = row.names(subset(ciliated_cds_pr_test_res, q_value < 0.05))
# ä»pr_deg_idsä¸­æŒ‘é€‰å‡ ä¸ªåŸºå› ï¼Œç„¶åå¯è§†åŒ–
plot_cells(cds, genes=c("hlh-4", "gcy-8", "dac-1", "oig-8"),
           show_trajectory_graph=FALSE,
           label_cell_groups=FALSE,
           label_leaves=FALSE)

# å¯ä»¥å°†è¿™äº›åœ¨è½¨è¿¹ä¸Šå˜åŒ–çš„pr_deg_idsç»§ç»­åˆ†ä¸ºå°æ¨¡å—
gene_module_df = monocle3:::find_gene_modules(cds[pr_deg_ids,], resolution=c(0,10^seq(-6,-1)))

cell_group_df = tibble::tibble(cell=row.names(colData(cds)), cell_group=colData(cds)$cell.type)
agg_mat = aggregate_gene_expression(cds, gene_module_df, cell_group_df)
row.names(agg_mat) = stringr::str_c("Module ", row.names(agg_mat))
pheatmap::pheatmap(agg_mat,
                    scale="column", clustering_method="ward.D2")

plot_cells(cds,
           genes=gene_module_df %>% filter(module %in% c(29,20, 11,22)),
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE)
```

**å¦å¤–è¿˜æœ‰ä¸€ç§æ–¹æ³•ï¼š**

å…ˆç”»å®Œæ•´çš„è½¨è¿¹å›¾ï¼š

```r
plot_cells(cds, show_trajectory_graph=FALSE)
```

ç„¶åæŒ‘æŸä¸€ä¸ªç»†èƒäºšç¾¤å¯¹åº”çš„ä¸€æ®µè½¨è¿¹ï¼š

```r
# å‡è®¾è¿™é‡ŒAFDç»†èƒå¯¹åº”ä¸€æ®µ22ã€28ã€35ç»„æˆçš„è½¨è¿¹çº¿
AFD_genes = c("gcy-8", "dac-1", "oig-8")
AFD_lineage_cds = cds[rowData(cds)$gene_short_name %in% AFD_genes,
                      clusters(cds) %in% c(22, 28, 35)]

plot_genes_in_pseudotime(AFD_lineage_cds,
                         color_cells_by="embryo.time.bin",
                         min_expr=0.5)
```

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2019-08-28-143429.png)

å°±çŸ¥é“dac-1åŸºå› çš„æ¿€æ´»å‘ç”Ÿåœ¨å…¶ä»–ä¸¤ä¸ªåŸºå› ä¹‹å‰

### å¦‚æœä½¿ç”¨ç‰ˆæœ¬2...

> ç‰ˆæœ¬2çš„éƒ¨åˆ†åº”è¯¥æ˜¯è¦æ–°å†™ä¸€ç¯‡çš„ï¼ˆåæœŸç²˜ä¸Šç‰ˆæœ¬2çš„é“¾æ¥ï¼‰

é¦–å…ˆåˆ›å»ºå¯¹è±¡ï¼š

* éœ€è¦æŒ‡å®š`exprs`ã€`phenoData`ã€`featureData`ï¼Œåä¸¤ä¸ªéƒ½è¦æ˜¯`AnnotatedDataFrame`å¯¹è±¡ï¼ˆç›®çš„æ˜¯ä¸è®©ç”¨æˆ·éšä¾¿ä¿®æ”¹ï¼Œçœ‹æ¥è¿˜æ˜¯ç‰ˆæœ¬3æ¯”è¾ƒäººæ€§åŒ–ï¼‰ï¼Œç„¶åä¾ç„¶æ˜¯è¦ä¸è¡¨è¾¾çŸ©é˜µè¡Œã€åˆ—å¯¹åº”

```r
#do not run
HSMM_expr_matrix <- read.table("fpkm_matrix.txt")
HSMM_sample_sheet <- read.delim("cell_sample_sheet.txt")
HSMM_gene_annotation <- read.delim("gene_annotations.txt")

pd <- new("AnnotatedDataFrame", data = HSMM_sample_sheet)
fd <- new("AnnotatedDataFrame", data = HSMM_gene_annotation)
HSMM <- newCellDataSet(as.matrix(HSMM_expr_matrix),
    phenoData = pd, featureData = fd)
```

### å†™åœ¨æœ€å

#### ä½“éªŒ

ä¸€å¥è¯å°±æ˜¯ï¼šæ»¡æ€€æœŸå¾…ï¼Œæ”¶è·æ»¡æ»¡ï¼Œçœ‹å¾—è´¹åŠ²

#### **ä¼˜ç‚¹**

* åˆ›å»ºå¯¹è±¡å˜å¾—æ–¹ä¾¿è®¸å¤šï¼Œä»åŸæ¥çš„`new()`å‡½æ•°åˆ›å»ºå¯¹è±¡åˆ°ç°åœ¨çš„æ•°æ®æ¡†ï¼Œå‡å°‘äº†å‡½æ•°çš„è®°å¿†
* æ®è¯´æ˜¯å·®å¼‚åˆ†æç®—æ³•è¿›è¡Œäº†æ›´æ–°ï¼Œå…·ä½“æ–°ä¸æ–°æœ‰å¾…æ¢ç´¢

#### **ç¼ºç‚¹**

* Monocle3çš„ç¤ºä¾‹æ•°æ®åšçš„ä¸æ˜¯å¾ˆèµ°å¿ƒï¼Œæ•°æ®ä¸‹è½½é€Ÿåº¦å—é™ã€‚è¿™ä¸€ç‚¹ä¸å¦‚Seuratåšçš„æ¸…æ™°æ˜äº†
* å¦å¤–å¾ˆå¤šåœ°æ–¹ä¸èƒ½é‡å¤å‡ºæ¥ï¼ˆæ¯”å¦‚ç›®å‰åŠ è½½çš„shinnyåŠŸèƒ½è¿˜ä¸å®Œå–„ï¼Œä¼šå‡ºç°æ— å“åº”çš„çŠ¶æ€ï¼›æ–‡æ¡£è¿˜æœ‰ä¸€äº›å°é”™è¯¯ï¼‰
* è®¸å¤šä»£ç æ“ä½œè¿‡ç¨‹å¤æ‚
* ç‰ˆæœ¬2ã€3ä¹‹é—´ä»£æ²Ÿå¾ˆå¤§ï¼Œåƒæ˜¯å®Œå…¨ä¸åŒçš„ä¸¤ä¸ªç‰ˆæœ¬ï¼Œè€Œä¸æ˜¯åƒSeurat2ã€3å‡ ä¸ªå‚æ•°ã€å‡½æ•°çš„æ”¹å˜
* å¯è§†åŒ–çš„ç»“æœå¥½åƒæ¯”ç‰ˆæœ¬2å°‘äº†ä¸€äº›ï¼Œä¸çŸ¥é“è¿™éƒ¨åˆ†æ˜¯æ²¡æœ‰æ”¹å˜è¿˜æ˜¯è¢«æ•´åˆäº†
